
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../book-4/">
      
      
        <link rel="next" href="../book-6/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.2">
    
    
      
        <title>6-5. 细胞通讯分析 - Bioinformatic Tutorial (Daxing Research)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7bf56d0a.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#43" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-header__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatic Tutorial (Daxing Research)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              6-5. 细胞通讯分析
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
            </label>
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-nav__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Bioinformatic Tutorial (Daxing Research)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        写在前面
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          基础环境配置
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          基础环境配置
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_1/" class="md-nav__link">
        配置1. Python环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_2/" class="md-nav__link">
        配置2. Linux环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_3/" class="md-nav__link">
        配置3. R语言环境
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          RNA-seq上游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq上游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_1/" class="md-nav__link">
        1-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_2/" class="md-nav__link">
        1-2. 上游分析全教程
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          RNA-seq下游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq下游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_1/" class="md-nav__link">
        2-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_2/" class="md-nav__link">
        2-2. 差异表达基因(DEG)分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_3/" class="md-nav__link">
        2-3. 通路富集分析(GSEA)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          单细胞样本对齐
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          单细胞样本对齐
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_1/" class="md-nav__link">
        3-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_2/" class="md-nav__link">
        3-2. 数据预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_3/" class="md-nav__link">
        3-3. 模型准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_4/" class="md-nav__link">
        3-4. 模型训练
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_5/" class="md-nav__link">
        3-5. 对齐细胞
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          MOFA单细胞多组学因子分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          MOFA单细胞多组学因子分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_1/" class="md-nav__link">
        4-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_2/" class="md-nav__link">
        4-2. 单细胞样本对齐
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_3/" class="md-nav__link">
        4-3. scRNA-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_4/" class="md-nav__link">
        4-4. scATAC-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_5/" class="md-nav__link">
        4-5. MOFA模型构建
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_6/" class="md-nav__link">
        4-6. MOFA下游分析-R语言环节
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_7/" class="md-nav__link">
        4-7. MOFA下游分析-Python环节
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          TMT蛋白组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          TMT蛋白组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_1/" class="md-nav__link">
        5-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_2/" class="md-nav__link">
        5-2. TMT蛋白组学上游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_3/" class="md-nav__link">
        5-3. TMT蛋白组学下游分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" checked>
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          多组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          多组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-1/" class="md-nav__link">
        6-1. 数据准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-2/" class="md-nav__link">
        6-2. GLUE多组学整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-3/" class="md-nav__link">
        6-3. 细胞亚群分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-4/" class="md-nav__link">
        6-4. 差异表达分析
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          6-5. 细胞通讯分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        6-5. 细胞通讯分析
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#43" class="md-nav__link">
    4.3 细胞通讯分析
  </a>
  
    <nav class="md-nav" aria-label="4.3 细胞通讯分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#431" class="md-nav__link">
    4.3.1 细胞通讯分析数据准备
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#432" class="md-nav__link">
    4.3.2 细胞通讯分析
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#433-" class="md-nav__link">
    4.3.3 细胞通讯可视化-细胞关系
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#434-" class="md-nav__link">
    4.3.4 细胞通讯可视化-受配体关系
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-6/" class="md-nav__link">
        6-6. 动态调控分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-7/" class="md-nav__link">
        6-7. 转录因子网络
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          Python可视化
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Python可视化
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_1/" class="md-nav__link">
        1-1. 火山图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_2/" class="md-nav__link">
        1-2. 箱线图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_3/" class="md-nav__link">
        1-3. 热图(聚类)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          视网膜母细胞瘤分析教程
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          视网膜母细胞瘤分析教程
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/rb_tutorial/" class="md-nav__link">
        Base
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_1_preprocess/" class="md-nav__link">
        教程1: 单细胞上游结果预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_2_filtered_qc/" class="md-nav__link">
        教程2: 质量控制与过滤
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_3_glue/" class="md-nav__link">
        教程3: scRNA-seq与scATAC-seq整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_4_scbasset/" class="md-nav__link">
        教程4: Motif活性分数计算
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_5_analysis1/" class="md-nav__link">
        教程5: 综合分析1
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_6_analysis2/" class="md-nav__link">
        教程6: 综合分析1-Chromatin activity和Motif score评估
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_7_STER_cell/" class="md-nav__link">
        教程7: STER cell下游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_8_kineic_photo/" class="md-nav__link">
        教程8: 转录动力学分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_9_cellphonedb/" class="md-nav__link">
        教程9: 细胞通讯分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#43" class="md-nav__link">
    4.3 细胞通讯分析
  </a>
  
    <nav class="md-nav" aria-label="4.3 细胞通讯分析">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#431" class="md-nav__link">
    4.3.1 细胞通讯分析数据准备
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#432" class="md-nav__link">
    4.3.2 细胞通讯分析
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#433-" class="md-nav__link">
    4.3.3 细胞通讯可视化-细胞关系
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#434-" class="md-nav__link">
    4.3.4 细胞通讯可视化-受配体关系
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


  <h1>6-5. 细胞通讯分析</h1>

<h3 id="43">4.3 细胞通讯分析</h3>
<p>在分析完阳性淋巴结跟阴性淋巴结的差异表达情况后，我们还有一种很直观的思路就是，我们可以比较阳性淋巴结跟阴性淋巴结中，这5种不同的B细胞之间的通讯，是否也发生了相应的变化。</p>
<p>在这里，我们用到的一个新的分析方法，叫做CellphoneDB细胞通讯分析，顾名思义，细胞之间打电话，相关包的介绍我们可以参照官方说明文档，这里不做过多解释，那么，在我们的淋巴结中，该如何应用这个分析呢？</p>
<h4 id="431">4.3.1 细胞通讯分析数据准备</h4>
<p>我们首先还是导入包以及B细胞数据</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="linenos" data-linenos=" 2 "></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="linenos" data-linenos=" 3 "></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="linenos" data-linenos=" 4 "></span><span class="kn">import</span> <span class="nn">anndata</span>
<span class="linenos" data-linenos=" 5 "></span><span class="kn">import</span> <span class="nn">os</span>
<span class="linenos" data-linenos=" 6 "></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="linenos" data-linenos=" 7 "></span><span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
<span class="linenos" data-linenos=" 9 "></span>
<span class="linenos" data-linenos="10 "></span><span class="n">adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;B_cell_anno_new.h5ad&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="11 "></span><span class="n">adata</span>
</code></pre></div>
<p>在这里，我们将要用到的是原始表达矩阵，也就是没有归一化的数据，所以我们采用GLUE分析中产生的中间文件<code>rna-emb.h5ad</code>来做分析，首先是B细胞中的阳性淋巴结</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">adata_raw</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;../glue_true/rna-emb.h5ad&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">adata_raw</span><span class="o">=</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">to_adata</span><span class="p">()[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">adata_raw</span><span class="o">=</span><span class="n">adata_raw</span><span class="p">[</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Pos&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos=" 4 "></span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_raw</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">adata_raw</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">X</span><span class="p">,</span><span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos=" 9 "></span>                          <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
<span class="linenos" data-linenos="10 "></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;Bcell_pos_raw.h5ad&#39;</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="11 "></span><span class="n">adata_raw</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Bcell_pos_counts.tsv&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="12 "></span>
<span class="linenos" data-linenos="13 "></span>
<span class="linenos" data-linenos="14 "></span><span class="n">df_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Cell&#39;</span><span class="p">:</span><span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="15 "></span>                             <span class="s1">&#39;cell_type&#39;</span><span class="p">:[</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata_raw</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;B_celltype&#39;</span><span class="p">]]</span>
<span class="linenos" data-linenos="16 "></span>                            <span class="p">})</span>
<span class="linenos" data-linenos="17 "></span><span class="n">df_meta</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Cell&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="18 "></span><span class="n">df_meta</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Bcell_pos_meta.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span>
<span class="linenos" data-linenos="20 "></span><span class="n">adata_raw</span>
</code></pre></div>
<p>输出结果</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>AnnData<span class="w"> </span>object<span class="w"> </span>with<span class="w"> </span>n_obs<span class="w"> </span>×<span class="w"> </span><span class="nv">n_vars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2684</span><span class="w"> </span>×<span class="w"> </span><span class="m">12951</span>
</code></pre></div>
<p>通过运行该代码，我们一共得到了<code>Bcell_pos_counts.tsv</code>与<code>Bcell_pos_meta.tsv</code>两个文件，同理我们对阴性淋巴结进行相同的数据准备，将得到<code>Bcell_neg_counts.tsv</code>与<code>Bcell_neg_meta.tsv</code>两个文件。</p>
<h4 id="432">4.3.2 细胞通讯分析</h4>
<p>我们在终端<code>terminal</code>中输入以下代码，等待1个小时左右便能得到细胞通讯分析的运行结果了</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>cellphonedb<span class="w"> </span>method<span class="w"> </span>statistical_analysis<span class="w"> </span>/content/Bcell_neg_meta.tsv<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="2 "></span>/content/Bcell_neg_counts.tsv<span class="w"> </span>--output-path<span class="w"> </span>/content/Bcell_neg<span class="w"> </span>--counts-data<span class="w"> </span>hgnc_symbol
<span class="linenos" data-linenos="3 "></span>
<span class="linenos" data-linenos="4 "></span>cellphonedb<span class="w"> </span>method<span class="w"> </span>statistical_analysis<span class="w"> </span>/content/Bcell_pos_meta.tsv<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="5 "></span>/content/Bcell_pos_counts.tsv<span class="w"> </span>--output-path<span class="w"> </span>/content/Bcell_pos<span class="w"> </span>--counts-data<span class="w"> </span>hgnc_symbol
</code></pre></div>
<p>输出结果</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:07:53<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Initializing<span class="w"> </span>SqlAlchemy<span class="w"> </span>CellPhoneDB<span class="w"> </span>Core
<span class="linenos" data-linenos=" 2 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:07:53<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Using<span class="w"> </span>custom<span class="w"> </span>database<span class="w"> </span>at<span class="w"> </span>/root/.cpdb/releases/v4.0.0/cellphone.db
<span class="linenos" data-linenos=" 3 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:07:53<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>cpdb_statistical_analysis_local_method_launcher
<span class="linenos" data-linenos=" 4 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:07:53<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_set_paths
<span class="linenos" data-linenos=" 5 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:07:53<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_load_meta_counts
<span class="linenos" data-linenos=" 6 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:08:00<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_check_counts_data
<span class="linenos" data-linenos=" 7 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:00<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>cpdb_statistical_analysis_launcher
<span class="linenos" data-linenos=" 8 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:00<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_counts_validations
<span class="linenos" data-linenos=" 9 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:00<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>get_interactions_genes_complex
<span class="linenos" data-linenos="10 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:00<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span><span class="o">[</span>Cluster<span class="w"> </span>Statistical<span class="w"> </span>Analysis<span class="o">]</span><span class="w"> </span>Threshold:0.1<span class="w"> </span>Iterations:1000<span class="w"> </span>Debug-seed:-1<span class="w"> </span>Threads:4<span class="w"> </span>Precision:3
<span class="linenos" data-linenos="11 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:02<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Running<span class="w"> </span>Real<span class="w"> </span>Analysis
<span class="linenos" data-linenos="12 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:08:02<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Running<span class="w"> </span>Statistical<span class="w"> </span>Analysis
<span class="linenos" data-linenos="13 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:47<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Building<span class="w"> </span>Pvalues<span class="w"> </span>result
<span class="linenos" data-linenos="14 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:48<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Building<span class="w"> </span>results
<span class="linenos" data-linenos="15 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:51<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Initializing<span class="w"> </span>SqlAlchemy<span class="w"> </span>CellPhoneDB<span class="w"> </span>Core
<span class="linenos" data-linenos="16 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:51<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Using<span class="w"> </span>custom<span class="w"> </span>database<span class="w"> </span>at<span class="w"> </span>/root/.cpdb/releases/v4.0.0/cellphone.db
<span class="linenos" data-linenos="17 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:39:51<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>cpdb_statistical_analysis_local_method_launcher
<span class="linenos" data-linenos="18 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:39:51<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_set_paths
<span class="linenos" data-linenos="19 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:39:51<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_load_meta_counts
<span class="linenos" data-linenos="20 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>APP<span class="o">][</span><span class="m">09</span>/10/22-18:39:57<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_check_counts_data
<span class="linenos" data-linenos="21 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:57<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>cpdb_statistical_analysis_launcher
<span class="linenos" data-linenos="22 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:57<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>_counts_validations
<span class="linenos" data-linenos="23 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:57<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Launching<span class="w"> </span>Method<span class="w"> </span>get_interactions_genes_complex
<span class="linenos" data-linenos="24 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:57<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span><span class="o">[</span>Cluster<span class="w"> </span>Statistical<span class="w"> </span>Analysis<span class="o">]</span><span class="w"> </span>Threshold:0.1<span class="w"> </span>Iterations:1000<span class="w"> </span>Debug-seed:-1<span class="w"> </span>Threads:4<span class="w"> </span>Precision:3
<span class="linenos" data-linenos="25 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:59<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Running<span class="w"> </span>Real<span class="w"> </span>Analysis
<span class="linenos" data-linenos="26 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-18:39:59<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Running<span class="w"> </span>Statistical<span class="w"> </span>Analysis
<span class="linenos" data-linenos="27 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-19:06:40<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Building<span class="w"> </span>Pvalues<span class="w"> </span>result
<span class="linenos" data-linenos="28 "></span><span class="o">[</span><span class="w"> </span><span class="o">][</span>CORE<span class="o">][</span><span class="m">09</span>/10/22-19:06:40<span class="o">][</span>INFO<span class="o">]</span><span class="w"> </span>Building<span class="w"> </span>results
</code></pre></div>
<p>我们将细胞通讯的结果分别存放到了<code>Bcell_pos</code>,<code>Bcell_neg</code>两个文件夹中，我们这里演示如何对阳性淋巴结(Pos)进行分析</p>
<h4 id="433-">4.3.3 细胞通讯可视化-细胞关系</h4>
<p>首先依然是导入基础的包</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="linenos" data-linenos=" 2 "></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="linenos" data-linenos=" 3 "></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="linenos" data-linenos=" 4 "></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="linenos" data-linenos=" 5 "></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="linenos" data-linenos=" 6 "></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="linenos" data-linenos=" 7 "></span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span>
<span class="linenos" data-linenos="11 "></span><span class="n">sc_color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#7CBB5F&#39;</span><span class="p">,</span><span class="s1">&#39;#368650&#39;</span><span class="p">,</span><span class="s1">&#39;#A499CC&#39;</span><span class="p">,</span><span class="s1">&#39;#5E4D9A&#39;</span><span class="p">,</span><span class="s1">&#39;#78C2ED&#39;</span><span class="p">,</span><span class="s1">&#39;#866017&#39;</span><span class="p">,</span><span class="s1">&#39;#9F987F&#39;</span><span class="p">,</span> <span class="s1">&#39;#E0DFED&#39;</span><span class="p">,</span> <span class="s1">&#39;#EF7B77&#39;</span><span class="p">,</span> <span class="s1">&#39;#279AD7&#39;</span><span class="p">,</span>
<span class="linenos" data-linenos="12 "></span> <span class="s1">&#39;#F0EEF0&#39;</span><span class="p">,</span> <span class="s1">&#39;#1F577B&#39;</span><span class="p">,</span> <span class="s1">&#39;#A56BA7&#39;</span><span class="p">,</span> <span class="s1">&#39;#E0A7C8&#39;</span><span class="p">,</span> <span class="s1">&#39;#E069A6&#39;</span><span class="p">,</span> <span class="s1">&#39;#941456&#39;</span><span class="p">,</span> <span class="s1">&#39;#FCBC10&#39;</span><span class="p">,</span> <span class="s1">&#39;#EAEFC5&#39;</span><span class="p">,</span> <span class="s1">&#39;#01A0A7&#39;</span><span class="p">,</span> <span class="s1">&#39;#75C8CC&#39;</span><span class="p">,</span> 
<span class="linenos" data-linenos="13 "></span><span class="s1">&#39;#F0D7BC&#39;</span><span class="p">,</span> <span class="s1">&#39;#D5B26C&#39;</span><span class="p">,</span> <span class="s1">&#39;#D5DA48&#39;</span><span class="p">,</span> <span class="s1">&#39;#B6B812&#39;</span><span class="p">,</span><span class="s1">&#39;#9DC3C3&#39;</span><span class="p">,</span> <span class="s1">&#39;#A89C92&#39;</span><span class="p">,</span> <span class="s1">&#39;#FEE00C&#39;</span><span class="p">,</span><span class="s1">&#39;#FEF2A1&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos="14 "></span>
<span class="linenos" data-linenos="15 "></span><span class="n">type_color_B</span><span class="o">=</span><span class="p">{</span>
<span class="linenos" data-linenos="16 "></span>    <span class="s1">&#39;naïve B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">13</span><span class="p">],</span>
<span class="linenos" data-linenos="17 "></span>    <span class="s1">&#39;Follicular B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">20</span><span class="p">],</span>
<span class="linenos" data-linenos="18 "></span>    <span class="s1">&#39;Pre-B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">14</span><span class="p">],</span>
<span class="linenos" data-linenos="19 "></span>    <span class="s1">&#39;activated B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">15</span><span class="p">],</span>
<span class="linenos" data-linenos="20 "></span>    <span class="s1">&#39;IgA+ Plasma B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">8</span><span class="p">],</span>
<span class="linenos" data-linenos="21 "></span>    <span class="s1">&#39;IgG+ Plasma B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">12</span><span class="p">],</span>
<span class="linenos" data-linenos="22 "></span>    <span class="s1">&#39;Plasma B cells&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">12</span><span class="p">],</span>
<span class="linenos" data-linenos="23 "></span>    <span class="s1">&#39;B-cell&#39;</span><span class="p">:</span><span class="n">sc_color</span><span class="p">[</span><span class="mi">13</span><span class="p">],</span>
<span class="linenos" data-linenos="24 "></span>    <span class="s1">&#39;B-cell (MALAT1+)&#39;</span><span class="p">:</span><span class="s1">&#39;#a3007d&#39;</span><span class="p">,</span>
<span class="linenos" data-linenos="25 "></span>
<span class="linenos" data-linenos="26 "></span><span class="p">}</span>
</code></pre></div>
<p>接着我们读取需要用到的文件</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">sig_means</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../cellphonedb/Bcell_pos/significant_means.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">sig_means</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<figure>
<p><img alt="Image title" src="../book-5.assets/image-20221019014049480-6120013.png" width="300" /></p>
</figure>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">de</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../cellphonedb/Bcell_pos/deconvoluted.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">de</span><span class="o">.</span><span class="n">columns</span><span class="p">[:]</span>
</code></pre></div>
<p>输出结果</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>Index<span class="o">([</span><span class="s1">&#39;gene_name&#39;</span>,<span class="w"> </span><span class="s1">&#39;uniprot&#39;</span>,<span class="w"> </span><span class="s1">&#39;is_complex&#39;</span>,<span class="w"> </span><span class="s1">&#39;protein_name&#39;</span>,<span class="w"> </span><span class="s1">&#39;complex_name&#39;</span>,
<span class="linenos" data-linenos="2 "></span><span class="w">       </span><span class="s1">&#39;id_cp_interaction&#39;</span>,<span class="w"> </span><span class="s1">&#39;B-cell (MALAT1+)&#39;</span>,<span class="w"> </span><span class="s1">&#39;Plasma B cells&#39;</span>,
<span class="linenos" data-linenos="3 "></span><span class="w">       </span><span class="s1">&#39;Pre-B cells&#39;</span>,<span class="w"> </span><span class="s1">&#39;activated B cells&#39;</span>,<span class="w"> </span><span class="s1">&#39;naïve B cells&#39;</span><span class="o">]</span>,
<span class="linenos" data-linenos="4 "></span><span class="w">      </span><span class="nv">dtype</span><span class="o">=</span><span class="s1">&#39;object&#39;</span><span class="o">)</span>
</code></pre></div>
<p>构建细胞通讯网络</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">G</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="linenos" data-linenos=" 2 "></span><span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">de</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">6</span><span class="p">:]:</span>
<span class="linenos" data-linenos=" 3 "></span>    <span class="n">test</span><span class="o">=</span><span class="n">sig_means</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">sig_means</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;\|&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="linenos" data-linenos=" 4 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">test</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
<span class="linenos" data-linenos=" 5 "></span>        <span class="k">if</span> <span class="n">test</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos=" 6 "></span>            <span class="n">a</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span>            <span class="n">G</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">weight</span><span class="o">=</span><span class="n">test</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="linenos" data-linenos=" 8 "></span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span> 
<span class="linenos" data-linenos="10 "></span>
<span class="linenos" data-linenos="11 "></span><span class="n">nodes</span><span class="o">=</span><span class="p">[]</span>
<span class="linenos" data-linenos="12 "></span><span class="n">G_degree</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()))</span>
<span class="linenos" data-linenos="13 "></span>
<span class="linenos" data-linenos="14 "></span>
<span class="linenos" data-linenos="15 "></span><span class="n">G_nodes_dict</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="16 "></span><span class="n">links</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="17 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
<span class="linenos" data-linenos="18 "></span>    <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G_nodes_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="linenos" data-linenos="19 "></span>        <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="20 "></span>    <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G_nodes_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="linenos" data-linenos="21 "></span>        <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="22 "></span>    <span class="n">links</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;target&quot;</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]})</span>
<span class="linenos" data-linenos="23 "></span>    <span class="n">weight</span><span class="o">=</span><span class="n">G</span><span class="o">.</span><span class="n">get_edge_data</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos="24 "></span>    <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">+=</span><span class="n">weight</span>
<span class="linenos" data-linenos="25 "></span>    <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">+=</span><span class="n">weight</span>
<span class="linenos" data-linenos="26 "></span>
<span class="linenos" data-linenos="27 "></span>
<span class="linenos" data-linenos="28 "></span><span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
<span class="linenos" data-linenos="29 "></span><span class="n">p</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<span class="linenos" data-linenos="30 "></span><span class="n">nodesize</span><span class="o">=</span><span class="p">[</span><span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">u</span><span class="p">]</span><span class="o">*</span><span class="mi">30</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span>
<span class="linenos" data-linenos="31 "></span><span class="n">nodecolos</span><span class="o">=</span><span class="p">[</span><span class="n">type_color_B</span><span class="p">[</span><span class="n">u</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span>
<span class="linenos" data-linenos="32 "></span><span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">nodelist</span><span class="o">=</span><span class="n">p</span><span class="p">,</span><span class="n">node_size</span><span class="o">=</span><span class="n">nodesize</span><span class="p">,</span><span class="n">node_color</span><span class="o">=</span><span class="n">nodecolos</span><span class="p">)</span><span class="c1">#cmap=nodecmap,</span>
<span class="linenos" data-linenos="33 "></span>                       <span class="c1">#node_color=list(p.values()))</span>
<span class="linenos" data-linenos="34 "></span>
<span class="linenos" data-linenos="35 "></span><span class="n">edgewidth</span> <span class="o">=</span> <span class="p">[</span><span class="n">G</span><span class="o">.</span><span class="n">get_edge_data</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">)[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">()]</span>
<span class="linenos" data-linenos="36 "></span><span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="n">edgewidth</span><span class="p">)</span>
<span class="linenos" data-linenos="37 "></span>
<span class="linenos" data-linenos="38 "></span>
<span class="linenos" data-linenos="39 "></span><span class="c1">#label_options = {&quot;ec&quot;: &quot;white&quot;, &quot;fc&quot;: &quot;white&quot;, &quot;alpha&quot;: 0.6}</span>
<span class="linenos" data-linenos="40 "></span><span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_labels</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,)</span> <span class="c1">#bbox=label_options)</span>
<span class="linenos" data-linenos="41 "></span><span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="42 "></span><span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="43 "></span><span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="44 "></span><span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mf">1.5</span><span class="p">)</span>
<span class="linenos" data-linenos="45 "></span><span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Pos Cellphone&#39;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="46 "></span><span class="c1">#plt.savefig(&quot;cellphonedb/pos_cellphone.png&quot;,dpi=300,bbox_inches = &#39;tight&#39;)</span>
</code></pre></div>
<figure>
<p><img alt="Image title" src="../book-5.assets/image-20221019014313452-6120017.png" width="300" /></p>
</figure>
<h4 id="434-">4.3.4 细胞通讯可视化-受配体关系</h4>
<p>首先依然是导入数据</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">sig_means1</span><span class="o">=</span><span class="n">sig_means</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">sig_means1</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">sig_means1</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos=" 3 "></span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">pvalue</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../cellphonedb/Bcell_pos/pvalues.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">pvalue</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">pvalue</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">p</span><span class="o">=</span><span class="n">pvalue</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">pvalue</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)]</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">p</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="linenos" data-linenos=" 8 "></span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">exp</span><span class="o">=</span><span class="n">sig_means1</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">sig_means1</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;\|&#39;</span><span class="p">)]</span>
<span class="linenos" data-linenos="10 "></span><span class="n">exp</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<p>紧接着，我们生成需要展示的受配体DataFrame</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">plot_data3</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pair&#39;</span><span class="p">,</span><span class="s1">&#39;means&#39;</span><span class="p">,</span><span class="s1">&#39;p-value&#39;</span><span class="p">,</span><span class="s1">&#39;Type&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">p_pair</span><span class="o">=</span><span class="n">exp</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos=" 3 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">exp</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())):</span>
<span class="linenos" data-linenos=" 4 "></span>    <span class="n">plot_data4</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos=" 5 "></span>    <span class="k">if</span> <span class="n">exp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">p_pair</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos=" 6 "></span>        <span class="k">continue</span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="n">plot_data4</span><span class="p">[</span><span class="s1">&#39;Pair&#39;</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">j</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">p_pair</span><span class="p">)]</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="n">plot_data4</span><span class="p">[</span><span class="s1">&#39;means&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">exp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">p_pair</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos=" 9 "></span>    <span class="n">plot_data4</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">p_pair</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;float&#39;</span><span class="p">))</span><span class="o">+</span><span class="mf">0.3</span>
<span class="linenos" data-linenos="10 "></span>    <span class="n">plot_data4</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">p_pair</span>
<span class="linenos" data-linenos="11 "></span>
<span class="linenos" data-linenos="12 "></span>    <span class="n">plot_data3</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_data3</span><span class="p">,</span><span class="n">plot_data4</span><span class="p">])</span>
<span class="linenos" data-linenos="13 "></span><span class="n">plot_data3</span><span class="o">=</span><span class="n">plot_data3</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="14 "></span>
<span class="linenos" data-linenos="15 "></span><span class="n">plot_data3</span><span class="p">[</span><span class="s1">&#39;means&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">plot_data3</span><span class="p">[</span><span class="s1">&#39;means&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos="16 "></span><span class="n">plot_data3</span><span class="o">=</span><span class="n">plot_data3</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;inf&#39;</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
<span class="linenos" data-linenos="17 "></span><span class="n">plot_data3</span><span class="o">=</span><span class="n">plot_data3</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s1">&#39;-inf&#39;</span><span class="p">),</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="linenos" data-linenos="18 "></span><span class="n">plot_data3</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<table>
<thead>
<tr>
<th align="right">Pair</th>
<th align="right">means</th>
<th align="right">p-value</th>
<th align="right">Type</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td align="right">0</td>
<td align="right">CCR10_CCL28</td>
<td align="right">-3.0</td>
<td align="right">0.3</td>
<td>B-cell (MALAT1+)|B-cell (MALAT1+)</td>
</tr>
<tr>
<td align="right">1</td>
<td align="right">CCR10_CCL28</td>
<td align="right">-3.0</td>
<td align="right">0.3</td>
<td>B-cell (MALAT1+)|Plasma B cells</td>
</tr>
<tr>
<td align="right">2</td>
<td align="right">CCR10_CCL28</td>
<td align="right">-3.0</td>
<td align="right">0.3</td>
<td>B-cell (MALAT1+)|Pre-B cells</td>
</tr>
<tr>
<td align="right">3</td>
<td align="right">CCR10_CCL28</td>
<td align="right">-3.0</td>
<td align="right">0.3</td>
<td>B-cell (MALAT1+)|activated B cells</td>
</tr>
<tr>
<td align="right">4</td>
<td align="right">CCR10_CCL28</td>
<td align="right">-3.0</td>
<td align="right">0.3</td>
<td>B-cell (MALAT1+)|naïve B cells</td>
</tr>
</tbody>
</table>
<p>基础绘图</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1">#pip install python_dotplot</span>
<span class="linenos" data-linenos=" 2 "></span><span class="kn">import</span> <span class="nn">dotplot</span>
<span class="linenos" data-linenos=" 3 "></span><span class="kn">import</span> <span class="nn">dotplot.utils</span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">new_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;item_key&#39;</span><span class="p">:</span> <span class="s1">&#39;Pair&#39;</span><span class="p">,</span><span class="s1">&#39;group_key&#39;</span><span class="p">:</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span><span class="s1">&#39;sizes_key&#39;</span><span class="p">:</span> <span class="s1">&#39;p-value&#39;</span><span class="p">,</span><span class="s1">&#39;color_key&#39;</span><span class="p">:</span> <span class="s1">&#39;means&#39;</span><span class="p">}</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">dp</span> <span class="o">=</span> <span class="n">dotplot</span><span class="o">.</span><span class="n">DotPlot</span><span class="o">.</span><span class="n">parse_from_tidy_data</span><span class="p">(</span><span class="n">plot_data3</span><span class="p">,</span> <span class="o">**</span><span class="n">new_keys</span><span class="p">)</span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">ax</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">size_factor</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Spectral_r&#39;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 7 "></span>             <span class="n">dot_title</span> <span class="o">=</span> <span class="s1">&#39;–log10(Pvalue)&#39;</span><span class="p">,</span> <span class="n">colorbar_title</span> <span class="o">=</span> <span class="s1">&#39;log2(means)&#39;</span><span class="p">,</span><span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">xlabs</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos="10 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="11 "></span><span class="n">ylabs</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos="12 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
</code></pre></div>
<figure>
<p><img alt="Image title" src="../book-5.assets/image-20221019014538577-6120020.png" width="300" /></p>
</figure>
<p>这图看起来不是很美观，我们稍微美化一下，将生成两张图，最后再用PS合并，首先是受体图</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1">#去除配体</span>
<span class="linenos" data-linenos=" 2 "></span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;Vil-&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xlabs</span><span class="p">]</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos=" 4 "></span><span class="c1">#绘制虚线</span>
<span class="linenos" data-linenos=" 5 "></span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ylabs</span><span class="p">]</span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos=" 7 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabs</span><span class="p">)):</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="k">if</span> <span class="n">i</span><span class="o">%</span><span class="mi">5</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos=" 9 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">i</span><span class="o">+</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">ymin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span><span class="c1">#上色        </span>
<span class="linenos" data-linenos="11 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span><span class="n">colors</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="12 "></span><span class="n">ax</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;cellphonedb/figure_pos_rec.png&quot;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span><span class="n">ax</span>
</code></pre></div>
<figure>
<p><img alt="Image title" src="../book-5.assets/image-20221019014718599-6120022.png" width="300" /></p>
</figure>
<p>其次是配体图</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">new_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;item_key&#39;</span><span class="p">:</span> <span class="s1">&#39;Pair&#39;</span><span class="p">,</span><span class="s1">&#39;group_key&#39;</span><span class="p">:</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span><span class="s1">&#39;sizes_key&#39;</span><span class="p">:</span> <span class="s1">&#39;p-value&#39;</span><span class="p">,</span><span class="s1">&#39;color_key&#39;</span><span class="p">:</span> <span class="s1">&#39;means&#39;</span><span class="p">}</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">dp</span> <span class="o">=</span> <span class="n">dotplot</span><span class="o">.</span><span class="n">DotPlot</span><span class="o">.</span><span class="n">parse_from_tidy_data</span><span class="p">(</span><span class="n">plot_data3</span><span class="p">,</span> <span class="o">**</span><span class="n">new_keys</span><span class="p">)</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">ax</span> <span class="o">=</span> <span class="n">dp</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">size_factor</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Spectral_r&#39;</span><span class="p">,</span>
<span class="linenos" data-linenos=" 4 "></span>             <span class="n">dot_title</span> <span class="o">=</span> <span class="s1">&#39;–log10(Pvalue)&#39;</span><span class="p">,</span> <span class="n">colorbar_title</span> <span class="o">=</span> <span class="s1">&#39;log2(means)&#39;</span><span class="p">,</span><span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos=" 5 "></span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">xlabs</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">ylabs</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span><span class="c1">#去除受体</span>
<span class="linenos" data-linenos="11 "></span><span class="p">[</span><span class="n">xlabs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;------&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabs</span><span class="p">))</span> <span class="k">if</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="p">)</span><span class="o">%</span><span class="mi">5</span><span class="o">!=</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="12 "></span><span class="p">[</span><span class="n">xlabs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="s1">&#39;------</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span><span class="n">xlabs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;p-value_&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabs</span><span class="p">))</span> <span class="k">if</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="p">)</span><span class="o">%</span><span class="mi">5</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="13 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="14 "></span><span class="c1">#绘制虚线</span>
<span class="linenos" data-linenos="15 "></span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span><span class="s1">&#39;  &#39;</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">get_text</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ylabs</span><span class="p">]</span>
<span class="linenos" data-linenos="16 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ylabs</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="17 "></span>
<span class="linenos" data-linenos="18 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabs</span><span class="p">)):</span>
<span class="linenos" data-linenos="19 "></span>    <span class="k">if</span> <span class="n">i</span><span class="o">%</span><span class="mi">5</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos="20 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">i</span><span class="o">+</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">ymin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="21 "></span><span class="c1">#上色</span>
<span class="linenos" data-linenos="22 "></span><span class="n">ax</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span><span class="n">colors</span><span class="o">=</span><span class="s1">&#39;#3C718D&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;major&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="23 "></span><span class="n">ax</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;cellphonedb/figure_pos_li.png&quot;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="24 "></span><span class="n">ax</span>
</code></pre></div>
<figure>
<p><img alt="Image title" src="../book-5.assets/image-20221019014912052-6120025.png" width="300" /></p>
</figure>





                

  

              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fc8c2696.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>