# 16.差异基因表达分析

[16. Differential gene expression analysis — Single-cell best practices](https://www.sc-best-practices.org/conditions/differential_gene_expression.html)

## 16.1. 背景

此章节是“注释”小节的深入延续，已经引入差异基因表达（DGE）作为一种工具，用于用细胞类型注释簇。在此，我们专注于差异基因表达测试的更高级应用场景，这些场景基于更复杂的实验设计，涉及一个或多个条件，如疾病、基因敲除或药物处理。在这些情境下，我们通常关心感兴趣的条件与某一参考条件之间的基因表达模式的差异的大小和显著性。这个参考条件可以是任何样本，但通常是健康样本。这种统计测试可以应用于任意组别，**但在单细胞RNA测序的情境下，它通常应用于细胞类型层面**。

差异基因表达分析试图推断在任何比较的组之间（通常是每种细胞类型的健康和病态之间）统计学上显著地过度或低表达的基因。

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression.jpg](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression.jpg)

这种分析的结果可能是影响并可能解释任何观察到的表型的基因集。这些基因集可以进一步仔细探究，例如，受影响的通路或诱导的细胞间通信变化。

差异基因表达分析通常返回每个比较基因和每个比较条件的log2倍数变化和调整后的p值。然后可以按p值对此列表进行排序，并进行更详细的探索。

一直为科学家们广泛运用的学生t检验是进行此类分析的一种方法。然而，它未能考虑到单细胞RNA测序的几个特殊性，例如来源于数据丢失的大量零值或需要复杂的实验设计。更具体地说，很少有足够的样本数量来在不跨基因汇总信息的情况下准确估计方差。此外，原始计数从未是给定样本中特定基因的表达的绝对测量。每个基因的实际读取数取决于文库准备的效率、非编码转录本的污染量和测序深度。因此，**学生t检验在单细胞RNA测序的灵敏度和特异性上都存在不足，更不用说实验设计的灵活性了。**

因此，差异基因表达测试是一个经典的生物信息学问题，已经有许多工具来解决这个问题。通常，这个问题目前从两个视角来解决，一是样本级视角，其中表达被聚合以创建“伪批量/pseudobulks”然后使用最初为批量表达样本设计的方法进行分析，如`edgeR`[[Robinson *et al.*, 2010](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id316)]或`DEseq2`[[Love *et al.*, 2014](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id317)]；另一个是细胞级视角，其中细胞使用诸如`MAST`[[Finak *et al.*, 2015](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id321)]或`glmmTMB`[[Brooks *et al.*, 2017](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id322)]的广义混合效应模型单独建模。DGE工具在数据集之间的共识和稳健性较低[[Das *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id323), [Wang *et al.*, 2019](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id318)]。如前所述，尽管单细胞数据包含技术噪声伪像，如数据丢失、零膨胀和高细胞间变异性[[Hicks *et al.*, 2017](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id319), [Luecken and Theis, 2019](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id337), [Vallejos *et al.*, 2017](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id320)]，但为批量[RNA](https://www.sc-best-practices.org/glossary.html#term-RNA)seq数据设计的方法与专门为scRNA-seq数据设计的方法相比表现得更好[[Das *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id323), [Jaakkola *et al.*, 2016](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id325), [Soneson and Robinson, 2018](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id324), [Squair *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id326)]。单细胞特异性方法往往特别容易将高表达的基因错误地标记为差异表达。

最近的一项研究强调了伪重复的问题，其中推断统计学应用于不是统计上独立的生物重复实验。未能考虑到重复实验（来自同一个体的细胞）的固有相关性会导致假阳性发现率（FDR）的增加[[Junttila *et al.*, 2022](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id330), [Squair *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id326), [Zimmerman *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id328)]。因此，**在进行DGE分析之前，应该应用批次效应校正或通过求和、求均值或每个个体的随机效应来聚合个体内特定细胞类型的表达值，即生成伪批量，以考虑样本内的相关性**[[Zimmerman *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id328)]。通常，伪批量方法（如`edgeR`、`DESeq2`或`Limma`[[Ritchie *et al.*, 2015](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id331)]）与求和聚合和混合模型（如设置随机效应的`MAST`）相比，都被认为优于那些不考虑它们的天真方法，如流行的Wilcoxon秩和检验或Seurat的[[Hao *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id338)]潜在模型[[Junttila *et al.*, 2022](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id330)]。

针对Zimmerman的论文，Murphy等人对Zimmerman的基准测试策略进行了严格的审查并对其进行了改进[[Murphy and Skene, 2022](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id329)]。**他们得出的结论是伪批量方法表现最佳，但是求和还是求均值的聚合方式哪种更好还需要进一步的研究。**

因此，在这个Notebook中，我们展示了如何使用两种工具进行差异表达分析：使用准似然测试的`edgeR`和使用随机效应的`MAST`。由于`edgeR`和`MAST`都是在R中实现的，我们使用`anndata2ri`包来同时在Python中处理AnnData对象和在R中处理SingeCellExperiment对象。

对于这两种方法，我们展示了两个使用案例：如何在完整数据上运行分析并对多种细胞类型进行测试，以及如何对特定的单一细胞类型进行测试。

## 16.2. 环境设置

```python
import warnings

warnings.filterwarnings("ignore")

import matplotlib.pyplot as plt
import seaborn as sns
import scanpy as sc
import pandas as pd
import numpy as np
import random
import sc_toolbox

import rpy2.rinterface_lib.callbacks
import anndata2ri
import logging

from rpy2.robjects import pandas2ri
from rpy2.robjects import r

sc.settings.verbosity = 0
rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)

pandas2ri.activate()
anndata2ri.activate()

%load_ext rpy2.ipython

```

```python
During startup - Warning message:
Setting LC_CTYPE failed, using "C"
```

```r
library(edgeR)
library(MAST)
```

## 16.3. 数据集准备

我们将使用Kang数据集，这是一个基于10x微滴的单细胞RNA测序外周血单个核细胞(PBMC)数据，来源于8名红斑狼疮患者在接受INF-β治疗前后的6小时(总共16个样本)[[Kang *et al.*, 2018](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id336)]。干扰素β以天然成纤维细胞或重组制剂(干扰素β-1a和干扰素β-1b)的形式使用，并具有与干扰素α相似的抗病毒和抗增殖特性。干扰素β已被批准用于治疗复发性-缓解性多发性硬化症和继发性进行性多发性硬化症。

首先，我们加载完整的数据集。

```r
adata = sc.read("kang.h5ad")
adata

```

```r
AnnData object with n_obs × n_vars = 24673 × 15706
    obs: 'nCount_RNA', 'nFeature_RNA', 'tsne1', 'tsne2', 'label', 'cluster', 'cell_type', 'replicate', 'nCount_SCT', 'nFeature_SCT', 'integrated_snn_res.0.4', 'seurat_clusters'
    var: 'name'
    obsm: 'X_pca', 'X_umap'
```

我们需要`.obs`中的`label`（其中包含条件标签）、`replicate`和`cell_type`列。

```r
adata.obs[:5]
```

|  | nCount_RNA | nFeature_RNA | tsne1 | tsne2 | label | cluster | cell_type | replicate | nCount_SCT | nFeature_SCT | integrated_snn_res.0.4 | seurat_clusters |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| index |  |  |  |  |  |  |  |  |  |  |  |  |
| AAACATACATTTCC-1 | 3017.0 | 877 | -27.640373 | 14.966629 | ctrl | 9 | CD14+ Monocytes | patient_1016 | 1704.0 | 711 | 1 | 1 |
| AAACATACCAGAAA-1 | 2481.0 | 713 | -27.493646 | 28.924885 | ctrl | 9 | CD14+ Monocytes | patient_1256 | 1614.0 | 662 | 1 | 1 |
| AAACATACCATGCA-1 | 703.0 | 337 | -10.468194 | -5.984389 | ctrl | 3 | CD4 T cells | patient_1488 | 908.0 | 337 | 6 | 6 |
| AAACATACCTCGCT-1 | 3420.0 | 850 | -24.367997 | 20.429285 | ctrl | 9 | CD14+ Monocytes | patient_1256 | 1738.0 | 653 | 1 | 1 |
| AAACATACCTGGTA-1 | 3158.0 | 1111 | 27.952170 | 24.159738 | ctrl | 4 | Dendritic cells | patient_1039 | 1857.0 | 928 | 12 | 12 |

我们需要使用原始计数，因此我们检查`.X`确实包含原始计数，并将它们放入我们的AnnData对象的`counts`层中。

```r
np.max(adata.X)
```

```r
3828.0
```

```r
adata.layers["counts"] = adata.X.copy()
```

8名对照组和8名疾病患者。

```python
print(len(adata[adata.obs["label"] == "ctrl"].obs["replicate"].cat.categories))
print(len(adata[adata.obs["label"] == "stim"].obs["replicate"].cat.categories))
```

```python
8
8
```

为了进行基本的质量控制，我们筛选出基因数少于200的细胞，以及在少于3个细胞中发现的基因。(忘记请回看质控内容)

```python
sc.pp.filter_cells(adata, min_genes=200)
sc.pp.filter_genes(adata, min_cells=3)
adata

```

```python
AnnData object with n_obs × n_vars = 24562 × 15701
    obs: 'nCount_RNA', 'nFeature_RNA', 'tsne1', 'tsne2', 'label', 'cluster', 'cell_type', 'replicate', 'nCount_SCT', 'nFeature_SCT', 'integrated_snn_res.0.4', 'seurat_clusters', 'n_genes'
    var: 'name', 'n_cells'
    obsm: 'X_pca', 'X_umap'
    layers: 'counts'

```

## 16.4. 伪批量/pseudobulks

edgeR是一个在R中实现的差异基因表达检测工具，最初是为批量基因表达数据设计的。它基于负二项分布、经验Bayes估计、精确测试、广义线性模型(GLMs)和拟似然测试，实现了广泛的统计方法。

如需更多详细信息，请参阅原始出版物[[Robinson et al., 2010](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id316)]。

在此，我们将使用准似然检验，因为它考虑了离散度估计的不确定性。相比之下，精确检验假设估计的离散度是真实值，这可能导致一些不准确性。此外，准似然广义线性模型(GLMs)在实验设计上更为灵活。

由于edgeR最初是作为批量数据的差异表达分析方法引入的，我们首先需要从我们的单细胞数据集中创建伪批量样本。对于每位患者，我们通过汇总每个亚群中的细胞并取该亚群内的平均基因表达来为每种细胞类型创建1个伪批量样本。

如果您正在处理的数据没有重复项，为了考虑患者的变异性，为每位患者创建多个（例如2-3个）伪批量可能是有益的。在这里，我们选择为每位患者创建1个伪批量，因为对于每位患者，我们有2个重复项，一个是对照组，另一个是刺激组。

我们强烈建议阅读这篇关于设计矩阵的指南：[https://f1000research.com/articles/9-1444](https://f1000research.com/articles/9-1444) 。

无论我们是否希望仅对少数细胞亚群进行分析并为它们每一个分别拟合一个模型，还是为它们所有人拟合一个模型，我们首先需要准备数据，定义一个创建伪批量的函数，并运行edgeR流程。首先，让我们准备数据。

由于我们需要为每个患者-条件组合创建伪批量，我们首先需要通过连接`replicate` 和 `label`来创建这样的列。

```python
adata.obs["sample"] = [
    f"{rep}_{l}" for rep, l in zip(adata.obs["replicate"], adata.obs["label"])
]
```

为了避免Python到R的转换问题，我们需要清理细胞类型名称，即用下划线替换空格并删除+符号。

```python
adata.obs["cell_type"] = [ct.replace(" ", "_") for ct in adata.obs["cell_type"]]
adata.obs["cell_type"] = [ct.replace("+", "") for ct in adata.obs["cell_type"]]
```

为了创建伪批次pseudobulk数据，我们需要确保元数据中的分类数据确实被设置为分类类型。

```python
adata.obs["replicate"] = adata.obs["replicate"].astype("category")
adata.obs["label"] = adata.obs["label"].astype("category")
adata.obs["sample"] = adata.obs["sample"].astype("category")
adata.obs["cell_type"] = adata.obs["cell_type"].astype("category")
```

现在，让我们定义一个函数，用于将单个细胞聚合成伪复制体：

- `aggregate_and_filter` 是一个函数，它从原始的单细胞AnnData对象中为指定的亚群为每个供体创建一个伪复制体的AnnData对象。在这里，我们还过滤掉那些对于指定群体有少于30个细胞的供体。
- 通过更改`replicates_per_patient`参数，可以为每个样本创建多个（n）伪复制体；然后将细胞分成n个大致相等的子集。

```python
NUM_OF_CELL_PER_DONOR = 30

def aggregate_and_filter(
    adata,
    cell_identity,
    donor_key="sample",
    condition_key="label",
    cell_identity_key="cell_type",
    obs_to_keep=[],  # which additional metadata to keep, e.g. gender, age, etc.
    replicates_per_patient=1,
):
    # subset adata to the given cell identity
    adata_cell_pop = adata[adata.obs[cell_identity_key] == cell_identity].copy()
    # check which donors to keep according to the number of cells specified with NUM_OF_CELL_PER_DONOR
    size_by_donor = adata_cell_pop.obs.groupby([donor_key]).size()
    donors_to_drop = [
        donor
        for donor in size_by_donor.index
        if size_by_donor[donor] <= NUM_OF_CELL_PER_DONOR
    ]
    if len(donors_to_drop) > 0:
        print("Dropping the following samples:")
        print(donors_to_drop)
    df = pd.DataFrame(columns=[*adata_cell_pop.var_names, *obs_to_keep])

    adata_cell_pop.obs[donor_key] = adata_cell_pop.obs[donor_key].astype("category")
    for i, donor in enumerate(donors := adata_cell_pop.obs[donor_key].cat.categories):
        print(f"\tProcessing donor {i+1} out of {len(donors)}...", end="\r")
        if donor not in donors_to_drop:
            adata_donor = adata_cell_pop[adata_cell_pop.obs[donor_key] == donor]
            # create replicates for each donor
            indices = list(adata_donor.obs_names)
            random.shuffle(indices)
            indices = np.array_split(np.array(indices), replicates_per_patient)
            for i, rep_idx in enumerate(indices):
                adata_replicate = adata_donor[rep_idx]
                # specify how to aggregate: sum gene expression for each gene for each donor and also keep the condition information
                agg_dict = {gene: "sum" for gene in adata_replicate.var_names}
                for obs in obs_to_keep:
                    agg_dict[obs] = "first"
                # create a df with all genes, donor and condition info
                df_donor = pd.DataFrame(adata_replicate.X.A)
                df_donor.index = adata_replicate.obs_names
                df_donor.columns = adata_replicate.var_names
                df_donor = df_donor.join(adata_replicate.obs[obs_to_keep])
                # aggregate
                df_donor = df_donor.groupby(donor_key).agg(agg_dict)
                df_donor[donor_key] = donor
                df.loc[f"donor_{donor}_{i}"] = df_donor.loc[donor]
    print("\n")
    # create AnnData object from the df
    adata_cell_pop = sc.AnnData(
        df[adata_cell_pop.var_names], obs=df.drop(columns=adata_cell_pop.var_names)
    )
    return adata_cell_pop
```

我们还需要定义一个单独的函数来拟合edgeR的广义线性模型(GLM)：

- `fit_model` 接受一个SingleCellExperiment对象作为输入，创建设计矩阵，并输出拟合的GLM。我们还输出类DGEList的edgeR对象以进行一些探索性数据分析(EDA)。

```r
fit_model <- function(adata_){
    # create an edgeR object with counts and grouping factor
    y <- DGEList(assay(adata_, "X"), group = colData(adata_)$label)
    # filter out genes with low counts
    print("Dimensions before subsetting:")
    print(dim(y))
    print("")
    keep <- filterByExpr(y)
    y <- y[keep, , keep.lib.sizes=FALSE]
    print("Dimensions after subsetting:")
    print(dim(y))
    print("")
    # normalize
    y <- calcNormFactors(y)
    # create a vector that is concatentation of condition and cell type that we will later use with contrasts
    group <- paste0(colData(adata_)$label, ".", colData(adata_)$cell_type)
    replicate <- colData(adata_)$replicate
    # create a design matrix: here we have multiple donors so also consider that in the design matrix
    design <- model.matrix(~ 0 + group + replicate)
    # estimate dispersion
    y <- estimateDisp(y, design = design)
    # fit the model
    fit <- glmQLFit(y, design)
    return(list("fit"=fit, "design"=design, "y"=y))
}
```

现在我们已经定义了所有所需的函数，所以我们可以继续创建伪批次数据。我们可能稍后想查看可用的元数据，因此我们将其保留在AnnData对象中。

```r
obs_to_keep = ["label", "cell_type", "replicate", "sample"]
```

我们需要将原始计数传递给edgeR。因此，我们将`.X`设置为`counts`层，以确保为原始计数创建伪复制品。

```python
adata.X = adata.layers["counts"].copy()
```

接下来，我们创建带有伪批量数据的AnnData对象。

```python
# process first cell type separately...
cell_type = adata.obs["cell_type"].cat.categories[0]
print(
    f'Processing {cell_type} (1 out of {len(adata.obs["cell_type"].cat.categories)})...'
)
adata_pb = aggregate_and_filter(adata, cell_type, obs_to_keep=obs_to_keep)
for i, cell_type in enumerate(adata.obs["cell_type"].cat.categories[1:]):
    print(
        f'Processing {cell_type} ({i+2} out of {len(adata.obs["cell_type"].cat.categories)})...'
    )
    adata_cell_type = aggregate_and_filter(adata, cell_type, obs_to_keep=obs_to_keep)
    adata_pb = adata_pb.concatenate(adata_cell_type)

```

```python
Processing B_cells (1 out of 8)...
Dropping the following samples:
['patient_1039_ctrl']
	Processing donor 16 out of 16...

Processing CD14_Monocytes (2 out of 8)...
	Processing donor 16 out of 16...

Processing CD4_T_cells (3 out of 8)...
	Processing donor 16 out of 16...

Processing CD8_T_cells (4 out of 8)...
Dropping the following samples:
['patient_101_ctrl', 'patient_1039_ctrl', 'patient_1039_stim', 'patient_107_ctrl', 'patient_107_stim', 'patient_1244_ctrl', 'patient_1244_stim']
	Processing donor 16 out of 16...

Processing Dendritic_cells (5 out of 8)...
Dropping the following samples:
['patient_1016_ctrl', 'patient_1016_stim', 'patient_101_ctrl', 'patient_1039_ctrl', 'patient_1039_stim', 'patient_107_ctrl', 'patient_107_stim']
	Processing donor 16 out of 16...

Processing FCGR3A_Monocytes (6 out of 8)...
Dropping the following samples:
['patient_1039_ctrl', 'patient_1039_stim', 'patient_107_ctrl', 'patient_107_stim', 'patient_1244_stim']
	Processing donor 16 out of 16...

Processing Megakaryocytes (7 out of 8)...
Dropping the following samples:
['patient_1015_ctrl', 'patient_1015_stim', 'patient_1016_ctrl', 'patient_101_ctrl', 'patient_1039_stim', 'patient_107_stim', 'patient_1244_ctrl', 'patient_1244_stim', 'patient_1256_ctrl', 'patient_1256_stim', 'patient_1488_ctrl']
	Processing donor 11 out of 11...

Processing NK_cells (8 out of 8)...
Dropping the following samples:
['patient_1039_ctrl', 'patient_1039_stim']
	Processing donor 16 out of 16...

```

差异基因表达结果的有效性高度依赖于在统计模型中捕获主要变异轴。对伪批量样本进行中间数据探索步骤，如主成分分析(PCA)或多维缩放(MDS)，可以识别变异的来源，从而指导构建相应的设计和对比矩阵来模拟数据[[Law *et al.*, 2020](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id333)]。

对于包括生物重复实验的实验，未能考虑到多种生物变异性的来源会导致假阳性发现率(FDR)的增加[[Thurman *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id332)],[[Lähnemann *et al.*, 2020](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id334)]。虽然增加每个个体的细胞数量可以提高精确度，但它对检测个体间差异的功效影响有限。因此，**增加统计功效的最佳方法是增加独立的实验样本数量**[[Zimmerman *et al.*, 2021](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id328)]。

由于我们的数据已经生成，我们不能进一步增加独立实验样本的数量。尽管如此，我们现在将探索我们的数据，以确定主要的变异轴，从而正确地生成我们的设计矩阵。

我们对创建的伪复制品进行非常基本的探索性数据分析，以检查是否有一些患者/伪复制品是我们需要排除的异常值，以免偏倚差异表达结果。我们将原始计数保存在`counts`层中，然后对计数进行归一化，并为归一化的伪复制品计数计算PCA坐标。

```python
adata_pb.layers['counts'] = adata_pb.X.copy()
```

```python
sc.pp.normalize_total(adata_pb, target_sum=1e6)
sc.pp.log1p(adata_pb)
sc.pp.pca(adata_pb)
```

接下来，我们在PCA图上查看创建的伪复制品（pseudo-replicates），并按所有可用的元数据进行着色，以查看是否有任何我们可能希望包含在设计矩阵中的混杂因素。我们还添加了`lib_size`和`log_lib_size`列，以检查库大小与PC组件之间是否存在相关性。

```python
adata_pb.obs["lib_size"] = np.sum(adata_pb.layers["counts"], axis=1)
adata_pb.obs["log_lib_size"] = np.log(adata_pb.obs["lib_size"])
```

```python
sc.pl.pca(adata_pb, color=adata_pb.obs, ncols=1, size=300)
```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_49_0.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_49_0.png)

在PCA图上，我们观察到细胞类型的分离以及刺激和未刺激细胞的分离。其他协变量（批次）似乎与PCA组件没有明显的相关性，因此我们没有将它们任何一个包含到我们的设计矩阵中。

如上所述，edgeR以原始计数为输入，因此在我们继续之前，我们将计数放回`.X`字段中。

```python
adata_pb.X = adata_pb.layers['counts'].copy()
```

### 16.4.1. 单一细胞群体

首先，我们展示如何为特定的细胞类型准备数据、构建设计矩阵并进行差异表达测试。

我们在数据的CD14+单核细胞子集上运行此流程，因为文献中显示在这个亚群中鉴定出了最多的差异表达基因。

```python
adata_mono = adata_pb[adata_pb.obs["cell_type"] == "CD14_Monocytes"]
adata_mono
```

```python
View of AnnData object with n_obs × n_vars = 16 × 15701
    obs: 'label', 'cell_type', 'replicate', 'sample', 'batch', 'lib_size', 'log_lib_size'
    uns: 'log1p', 'pca', 'label_colors', 'cell_type_colors', 'replicate_colors', 'sample_colors', 'batch_colors'
    obsm: 'X_pca'
    varm: 'PCs'
    layers: 'counts'
```

清理样本名称以使图表更为简洁

```python
adata_mono.obs_names = [
    name.split("_")[2] + "_" + name.split("_")[3] for name in adata_mono.obs_names
]
```

```python
%%time
%%R -i adata_mono
outs <-fit_model(adata_mono)
```

```python
[1] "Dimensions before subsetting:"
[1] 15701    16
[1] ""
[1] "Dimensions after subsetting:"
[1] 3709   16
[1] ""
CPU times: user 2.95 s, sys: 224 ms, total: 3.18 s
Wall time: 4.27 s
```

```r
%%R
fit <- outs$fit
y <- outs$y
```

由于我们在分析开始时并没有预先假设某个特定基因会上调或下调，因此我们需要可视化工具来理解差异基因表达(DGE)的结果。多维缩放(MDS)图允许我们进行高层次的概览。通常，我们期望来自不同条件的样本之间会有分离，如下图所示，我们的结果中也可以看到这种分离。

```r
%%R
plotMDS(y, col=ifelse(y$samples$group == "stim", "red", "blue"))

```

```python
Fontconfig warning: ignoring UTF-8: not a valid region tag

```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_61_1.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_61_1.png)

生物系数变异(Biological Coefficient of Variation, BCV)图展示了每个基因在生物组间的平均变异性作为平均表达的函数。例如，0.3的bcv表示基因在各组间的表达平均有30%的变异性。

**低丰度的基因通常表现出较大的BCV**，因为对于低丰度基因的读数测量更不确定。另一方面，平均表达量高的基因被更可靠地定量，因此它们通常具有较低的变异性，因此BCV较低。使用此图来检测异常基因或确定可能需要在设计矩阵中反映的其他实验因素。例如，图的右上角出现的一组平均表达量高且BCV高的基因可能标志着实验应激、污染等，特别是如果它们属于相似的基因家族。

BCV是离散度的平方根。标签间和常见BCV趋势之间的距离指示标签间（基因间）的离散度估计是否高度可变（即异质基因表达）。如果标签间的离散值非常异质，那么应用较少的调节来捕获异质性。这两条曲线上任意两点之间的距离反映了基因的离散度是如何缩小到常见趋势的。也就是说，它捕获了应用的离散度调节量。

> **生物系数变异（Biological Coefficient of Variation，BCV）图**是用来展示每个基因在不同生物组之间的平均变异性，其变异性是基于基因的平均表达水平。例如，如果一个基因的BCV为0.3，意味着在不同生物组之间，该基因的表达平均变异性为30%。
> 
> 
> BCV是离散度（dispersion）的平方根。在BCV图中，标签间（tagwise）和常见（common）BCV趋势之间的距离可以用来判断基因的离散度估计是否在不同基因之间变化很大，也就是是否存在基因表达异质性。如果标签间离散度值非常异质，说明不同基因的表达变异性很大，此时调节较少，以适应这种异质性。两个曲线上任意两点之间的距离表示对基因离散度进行了多少程度的调整，这也捕捉了离散度调节的量。
> 
> 在BCV图中，如果观察到一些低丰度的基因具有高BCV，但没有高丰度的基因具有高BCV，这表明对于高丰度基因来说，不太需要进一步的实验因素来纳入设计矩阵中，因为它们的表达变异性相对较低。
> 

在下面的BCV图中，我们观察到一些低丰度的基因具有高BCV，但没有高丰度的基因具有高BCV，这表明在设计矩阵中不应有任何进一步的实验考虑因素。

```r
%%R
plotBCV(y)
```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_63_0.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_63_0.png)

接下来，我们执行准似然检验以找出控制和刺激条件之间的差异表达基因。让我们检查设计矩阵有哪些列，以便指定正确的列进行测试。

```r
%%R
colnames(y$design)
```

```
[1] "groupctrl.CD14_Monocytes" "groupstim.CD14_Monocytes"
[3] "replicatepatient_107"     "replicatepatient_1015"
[5] "replicatepatient_1016"    "replicatepatient_1039"
[7] "replicatepatient_1244"    "replicatepatient_1256"
[9] "replicatepatient_1488"
```

```r
%%R -o tt
myContrast <- makeContrasts('groupstim.CD14_Monocytes-groupctrl.CD14_Monocytes', levels = y$design)
qlf <- glmQLFTest(fit, contrast=myContrast)
# get all of the DE genes and calculate Benjamini-Hochberg adjusted FDR
tt <- topTags(qlf, n = Inf)
tt <- tt$table
```

让我们查看该表格：对于未被edgeR过滤掉的每一个基因（在我们的情况下是3709个），该表格包含了差异表达测试的结果。该表格可以保存为`.csv`文件，以便后续使用和可视化。

在我们对完整数据进行分析后，我们将在本节末尾展示如何使用火山图。

```python
tt.shape
```

```python
(3709, 5)
```

```python
tt[:5]
```

|  | logFC | logCPM | F | PValue | FDR |
| --- | --- | --- | --- | --- | --- |
| HESX1 | 8.345536 | 6.773420 | 1281.013295 | 1.837373e-15 | 2.766927e-12 |
| CD38 | 7.126846 | 7.420668 | 1243.793133 | 2.266164e-15 | 2.766927e-12 |
| NT5C3A | 5.657050 | 8.327003 | 1218.102628 | 2.628780e-15 | 2.766927e-12 |
| SOCS1 | 4.388247 | 6.943768 | 1191.289806 | 3.079524e-15 | 2.766927e-12 |
| GMPR | 6.943484 | 7.031832 | 1159.601183 | 3.730018e-15 | 2.766927e-12 |

附带说明，使用`glmTreat`，我们还可以测试相对于指定的倍数变化，哪些基因在提供的系数或对比度下表现出差异表达。

```r
%%R
tr <- glmTreat(fit, contrast=myContrast, lfc=1.5)
print(head(topTags(tr)))

```

```r
Coefficient:  -1*groupctrl.CD14_Monocytes 1*groupstim.CD14_Monocytes
          logFC unshrunk.logFC    logCPM       PValue          FDR
HESX1  8.345536       8.702486  6.773420 2.108737e-14 3.535749e-11
CD38   7.126846       7.219184  7.420668 2.292459e-14 3.535749e-11
NT5C3A 5.657050       5.674640  8.327003 3.149544e-14 3.535749e-11
IL1RN  6.588583       6.596787 10.358946 4.249159e-14 3.535749e-11
GMPR   6.943484       7.047504  7.031832 4.766446e-14 3.535749e-11
DEFB1  6.654201       6.696759  8.067159 7.670147e-14 4.741429e-11
```

最后，我们可以看到我们有多少基因的FDR校正值小于0.01。

涂抹图（也称为MA图，M值与A值图）显示了基因的对数倍数变化与其平均丰度之间的关系。由于在低丰度范围内读数更为变异，我们通常观察到较高的logFC，从而导致较大的logFC估计。如果我们对logFC和平均logCPM值拟合一个loess曲线，该趋势应该围绕零居中。任何偏离这一点的情况都可能表明数据没有被正确地标准化。平均表达量大且绝对值中的logFC大的基因可以标记出对于进一步调查和跟进具有生物学意义的基因。

```r
%%R
plotSmear(qlf, de.tags = rownames(tt)[which(tt$FDR<0.01)])
```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_73_0.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_73_0.png)

### 16.4.2. ****多组分析****

接下来，我们展示如何为所有可用的细胞类型准备数据、构建设计矩阵并使用对照组在完整数据集上进行差异表达测试。

```r
%%time
%%R -i adata_pb
outs <-fit_model(adata_pb)
```

```
[1] "Dimensions before subsetting:"
[1] 15701    90
[1] ""
[1] "Dimensions after subsetting:"
[1] 2358   90
[1] ""
CPU times: user 11.9 s, sys: 39.7 ms, total: 12 s
Wall time: 12 s
```

```r
%%R
fit <- outs$fit
y <- outs$y
```

现在，我们使用对照组为每一种细胞类型执行准似然测试。由于在R循环内部没有直接的方法将表格从R移动到Python，所以我们为每种细胞类型手动获取结果。

```r
%%R -i adata_pb -o de_per_cell_type
de_per_cell_type <- list()
for (cell_type in unique(colData(adata_pb)$cell_type)) {
    print(cell_type)
    # create contrast for this cell type
    myContrast <- makeContrasts(paste0("groupstim.", cell_type, "-groupctrl.", cell_type), levels = y$design)
    # perform QLF test
    qlf <- glmQLFTest(fit, contrast=myContrast)
    # get all of the DE genes and calculate Benjamini-Hochberg adjusted FDR
    tt <- topTags(qlf, n = Inf)
    # save in the list with the results for all the cell types
    de_per_cell_type[[cell_type]] <- tt$table
}

```

```r
[1] "B_cells"
[1] "CD14_Monocytes"
[1] "CD4_T_cells"
[1] "CD8_T_cells"
[1] "Dendritic_cells"
[1] "FCGR3A_Monocytes"
[1] "NK_cells"

```

现在，让我们使用`sc_toolbox`包中的`sc_toolbox.tools.de_res_to_anndata`（来源于**[https://github.com/schillerlab/sc-toolbox](https://github.com/schillerlab/sc-toolbox)**）将其保存在原始`adata`对象的`.uns`中，该工具将结果保存为仿佛它们是使用scanpy的`rank_genes_groups()`函数创建的。以这种方式存储差异表达表的优势是我们现在也可以使用标准的scanpy绘图功能。我们将每种细胞类型的DEG表格保存为`.csv`格式，因为在细胞通讯（ cell-to-cell communication ）章节中我们稍后会需要它们。

```python
# get cell types that we ran the analysis for
cell_types = de_per_cell_type.keys()
# add the table to .uns for each cell type
for cell_type in cell_types:
    df = de_per_cell_type[cell_type]
    df["gene_symbol"] = df.index
    df["cell_type"] = cell_type
    sc_toolbox.tools.de_res_to_anndata(
        adata,
        df,
        groupby="cell_type",
        score_col="logCPM",
        pval_col="PValue",
        pval_adj_col="FDR",
        lfc_col="logFC",
        key_added="edgeR_" + cell_type,
    )
    df.to_csv(f"de_edgeR_{cell_type}.csv")

```

为了再次将表格获取为`pandas` DataFrame，我们使用`sc.get.rank_genes_groups_df`函数。

```
sc.get.rank_genes_groups_df(adata, group="CD14_Monocytes", key="edgeR_CD14_Monocytes")[
    :5
]
```

|  | names | scores | logfoldchanges | pvals | pvals_adj |
| --- | --- | --- | --- | --- | --- |
| 0 | FTH1 | 16.186098 | -0.572272 | 0.001589 | 0.002712 |
| 1 | MALAT1 | 16.025682 | 0.010049 | 0.877078 | 0.897245 |
| 2 | B2M | 15.468524 | 0.677266 | 0.0 | 0.0 |
| 3 | TMSB4X | 15.054697 | -0.217131 | 0.004663 | 0.007331 |
| 4 | FTL | 14.984937 | -0.041832 | 0.669951 | 0.710956 |

### 16.4.3. ****关于edgeR的注释****:

- 需要原始计数作为输入
- 需要来自单细胞实验的伪批量（pseudobulks）
- 如果单细胞实验中有多个供体，并且用户希望考虑**患者的变异性**，我们建议为每个患者创建2或3个伪复制品，并将患者信息纳入设计矩阵

## 16.5. ****单细胞特异性****

MAST使用两部分的广义线性模型来模拟单细胞基因表达。一部分模型描述了每个基因在细胞间的离散表达率，而另一部分模型描述了条件连续表达水平。

MAST框架使用两部分的广义线性模型来模拟单细胞基因表达。MAST的一个组件模拟了每个基因在细胞间的离散表达率，而另一个组件模拟了条件连续表达水平（基于基因被表达的条件）。更多详细信息，请阅读原始出版物[**[Finak *et al.*, 2015](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id321)**]。

MAST需要归一化的计数作为输入，所以我们首先取“计数”层，然后进行归一化步骤。

```
adata.X = adata.layers["counts"].copy()

```

```
sc.pp.normalize_total(adata, target_sum=1e6)
sc.pp.log1p(adata)

```

我们定义了一个小的辅助函数，用于处理R和Python之间的某些对象类型转换问题。我们还需要为每个亚群筛选出在少数细胞中表达的基因（在这种情况下为3），因为该模型需要能够估计每个基因的方差。

```
def prep_anndata(adata_):
    def fix_dtypes(adata_):
        df = pd.DataFrame(adata_.X.A, index=adata_.obs_names, columns=adata_.var_names)
        df = df.join(adata_.obs)
        return sc.AnnData(df[adata_.var_names], obs=df.drop(columns=adata_.var_names))

    adata_ = fix_dtypes(adata_)
    sc.pp.filter_genes(adata_, min_cells=3)
    return adata_
```

### 16.5.1. ****单一细胞群体****

与edgeR一样，可以使用完整数据集拟合模型，然后使用对照组对每个感兴趣的细胞类型进行测试，但在这个笔记本中，我们只展示了针对一个细胞类型的MAST-RE流程，即CD14单核细胞，以缩短运行时间。

```
adata_mono = adata[adata.obs["cell_type"] == "CD14_Monocytes"].copy()
adata_mono

```

```
AnnData object with n_obs × n_vars = 5696 × 15701
    obs: 'nCount_RNA', 'nFeature_RNA', 'tsne1', 'tsne2', 'label', 'cluster', 'cell_type', 'replicate', 'nCount_SCT', 'nFeature_SCT', 'integrated_snn_res.0.4', 'seurat_clusters', 'n_genes', 'sample'
    var: 'name', 'n_cells'
    uns: 'edgeR_B_cells', 'edgeR_CD14_Monocytes', 'edgeR_CD4_T_cells', 'edgeR_CD8_T_cells', 'edgeR_Dendritic_cells', 'edgeR_FCGR3A_Monocytes', 'edgeR_NK_cells', 'log1p'
    obsm: 'X_pca', 'X_umap'
    layers: 'counts'

```

```
sc.pp.filter_genes(adata_mono, min_cells=3)
adata_mono

```

```
AnnData object with n_obs × n_vars = 5696 × 12268
    obs: 'nCount_RNA', 'nFeature_RNA', 'tsne1', 'tsne2', 'label', 'cluster', 'cell_type', 'replicate', 'nCount_SCT', 'nFeature_SCT', 'integrated_snn_res.0.4', 'seurat_clusters', 'n_genes', 'sample'
    var: 'name', 'n_cells'
    uns: 'edgeR_B_cells', 'edgeR_CD14_Monocytes', 'edgeR_CD4_T_cells', 'edgeR_CD8_T_cells', 'edgeR_Dendritic_cells', 'edgeR_FCGR3A_Monocytes', 'edgeR_NK_cells', 'log1p'
    obsm: 'X_pca', 'X_umap'
    layers: 'counts'

```

接下来，我们按照上述方法过滤两个对象。

```
adata_mono = prep_anndata(adata_mono)
adata_mono

```

```
AnnData object with n_obs × n_vars = 5696 × 12268
    obs: 'nCount_RNA', 'nFeature_RNA', 'tsne1', 'tsne2', 'label', 'cluster', 'cell_type', 'replicate', 'nCount_SCT', 'nFeature_SCT', 'integrated_snn_res.0.4', 'seurat_clusters', 'n_genes', 'sample'
    var: 'n_cells'

```

与进行多次统计测试的情况一样，必须使用例如Benjamini-Hochberg校正[**[Luecken 和 Theis, 2019](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id337)**]，[**[Benjamini 和 Hochberg, 1995](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id335)**]来校正因条件变化而获得的DGE测试的p值。

与edgeR分析类似，我们定义了一个单独的函数用于分析：

- **`find_de_MAST_RE`**函数接受一个SingleCellExperiment对象作为输入，并运行带有RE流水线的MAST。该函数的输出是一个表格（在Python中是pandas DataFrame），其中包含分析的结果，例如每个基因的对数倍数变化、p值和FDR校正值。

```
adata_mono.obs["cell_type"] = [
    ct.replace(" ", "_") for ct in adata_mono.obs["cell_type"]
]
adata_mono.obs["cell_type"] = [
    ct.replace("+", "") for ct in adata_mono.obs["cell_type"]
]
```

```
%%R
find_de_MAST_RE <- function(adata_){
    # create a MAST object
    sca <- SceToSingleCellAssay(adata_, class = "SingleCellAssay")
    print("Dimensions before subsetting:")
    print(dim(sca))
    print("")
    # keep genes that are expressed in more than 10% of all cells
    sca <- sca[freq(sca)>0.1,]
    print("Dimensions after subsetting:")
    print(dim(sca))
    print("")
    # add a column to the data which contains scaled number of genes that are expressed in each cell
    cdr2 <- colSums(assay(sca)>0)
    colData(sca)$ngeneson <- scale(cdr2)
    # store the columns that we are interested in as factors
    label <- factor(colData(sca)$label)
    # set the reference level
    label <- relevel(label,"ctrl")
    colData(sca)$label <- label
    celltype <- factor(colData(sca)$cell_type)
    colData(sca)$celltype <- celltype
    # same for donors (which we need to model random effects)
    replicate <- factor(colData(sca)$replicate)
    colData(sca)$replicate <- replicate
    # create a group per condition-celltype combination
    colData(sca)$group <- paste0(colData(adata_)$label, ".", colData(adata_)$cell_type)
    colData(sca)$group <- factor(colData(sca)$group)
    # define and fit the model
    zlmCond <- zlm(formula = ~ngeneson + group + (1 | replicate),
                   sca=sca,
                   method='glmer',
                   ebayes=F,
                   strictConvergence=F,
                   fitArgsD=list(nAGQ = 0)) # to speed up calculations

    # perform likelihood-ratio test for the condition that we are interested in
    summaryCond <- summary(zlmCond, doLRT='groupstim.CD14_Monocytes')
    # get the table with log-fold changes and p-values
    summaryDt <- summaryCond$datatable
    result <- merge(summaryDt[contrast=='groupstim.CD14_Monocytes' & component=='H',.(primerid, `Pr(>Chisq)`)], # p-values
                     summaryDt[contrast=='groupstim.CD14_Monocytes' & component=='logFC', .(primerid, coef)],
                     by='primerid') # logFC coefficients
    # MAST uses natural logarithm so we convert the coefficients to log2 base to be comparable to edgeR
    result[,coef:=result[,coef]/log(2)]
    # do multiple testing correction
    result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]
    result = result[result$FDR<0.01,, drop=F]

    result <- stats::na.omit(as.data.frame(result))
    return(result)
}
```

我们为单核细胞运行该流程。

```
%%time
%%R -i adata_mono -o res
res <-find_de_MAST_RE(adata_mono)

```

```
[1] "Dimensions before subsetting:"
[1] 12268  5696
[1] ""
[1] "Dimensions after subsetting:"
[1] 1676 5696
[1] ""
CPU times: user 9min 35s, sys: 2.96 s, total: 9min 38s
Wall time: 9min 43s

```

让我康康（指结果

```
res[:5]

```

|  | primerid | Pr(>Chisq) | coef | FDR |
| --- | --- | --- | --- | --- |
| 1 | AAED1 | 8.410341e-19 | 0.709836 | 1.597106e-18 |
| 2 | ABI1 | 1.401688e-05 | 0.312797 | 1.824921e-05 |
| 3 | ABRACL | 3.920183e-06 | 0.418028 | 5.201004e-06 |
| 4 | ACADVL | 2.121110e-26 | -0.751305 | 4.656979e-26 |
| 5 | ACOT9 | 3.424174e-151 | 2.921133 | 2.689503e-150 |

我们像之前那样将结果存储在`.uns`中。请注意，我们不需要`score`列，所以我们只传递logFC。

```
res["gene_symbol"] = res["primerid"]
res["cell_type"] = "CD14_Monocytes"
sc_toolbox.tools.de_res_to_anndata(
    adata,
    res,
    groupby="cell_type",
    score_col="coef",
    pval_col="Pr(>Chisq)",
    pval_adj_col="FDR",
    lfc_col="coef",
    key_added="MAST_CD14_Monocytes",
)
```

```
adata_copy = adata.copy()
```

## 16.6. 可视化

我们可以使用热图和火山图来可视化结果。热图的每一行对应一个基因，每一列对应一个单细胞。颜色越亮，表示该基因在特定细胞中的表达越高。由于我们只绘制差异表达基因，我们希望在两种条件之间看到明显的表达差异。火山图经常用于可视化统计测试的结果，它们在x轴上显示表达的变化（对数倍数变化），在y轴上显示统计显著性（FDR校正的p值）。我们对FDR校正的p值小于0.01且对数倍数变化超过1.5的基因进行颜色编码。

在绘制热图之前，我们对数据进行归一化，以更好地看到两种条件之间的表达差异。

```
adata.X = adata.layers["counts"].copy()
sc.pp.normalize_total(adata, target_sum=1e6)
sc.pp.log1p(adata)
```

接下来，我们定义一个辅助绘图函数用于生成热图。

```
FDR = 0.01
LOG_FOLD_CHANGE = 1.5

def plot_heatmap(adata, group_key, group_name="cell_type", groupby="label"):
    cell_type = "_".join(group_key.split("_")[1:])
    res = sc.get.rank_genes_groups_df(adata, group=cell_type, key=group_key)
    res.index = res["names"].values
    res = res[
        (res["pvals_adj"] < FDR) & (abs(res["logfoldchanges"]) > LOG_FOLD_CHANGE)
    ].sort_values(by=["logfoldchanges"])
    print(f"Plotting {len(res)} genes...")
    markers = list(res.index)
    sc.pl.heatmap(
        adata[adata.obs[group_name] == cell_type].copy(),
        markers,
        groupby=groupby,
        swap_axes=True,
    )

```

最后，我们可以为CD14+单核细胞绘制展示edgeR和MAST with RE的差异表达基因（DEG）的热图。

```
plot_heatmap(adata, "edgeR_CD14_Monocytes")

```

```
Plotting 303 genes...

```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_115_1.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_115_1.png)

```
plot_heatmap(adata, "MAST_CD14_Monocytes")

```

```
Plotting 436 genes...

```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_116_1.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_116_1.png)

我们观察到，使用我们给定的调整后p值和对数倍数变化的阈值，MAST识别出了436个DEG，而edgeR识别出了303个基因。

接下来，我们定义火山图的辅助绘图函数。

```
FDR = 0.01
LOG_FOLD_CHANGE = 1.5

def volcano_plot(adata, group_key, group_name="cell_type", groupby="label", title=None):
    cell_type = "_".join(group_key.split("_")[1:])
    result = sc.get.rank_genes_groups_df(adata, group=cell_type, key=group_key).copy()
    result["-logQ"] = -np.log(result["pvals"].astype("float"))
    lowqval_de = result.loc[abs(result["logfoldchanges"]) > LOG_FOLD_CHANGE]
    other_de = result.loc[abs(result["logfoldchanges"]) <= LOG_FOLD_CHANGE]

    fig, ax = plt.subplots()
    sns.regplot(
        x=other_de["logfoldchanges"],
        y=other_de["-logQ"],
        fit_reg=False,
        scatter_kws={"s": 6},
    )
    sns.regplot(
        x=lowqval_de["logfoldchanges"],
        y=lowqval_de["-logQ"],
        fit_reg=False,
        scatter_kws={"s": 6},
    )
    ax.set_xlabel("log2 FC")
    ax.set_ylabel("-log Q-value")

    if title is None:
        title = group_key.replace("_", " ")
    plt.title(title)
    plt.show()

```

```
volcano_plot(adata, "MAST_CD14_Monocytes")

```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_120_0.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_120_0.png)

```
volcano_plot(adata, "edgeR_CD14_Monocytes")

```

![16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_121_0.png](16%20%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%20ceef081e2b1a40c380e8d814edc54f55/differential_gene_expression_121_0.png)

从热图和尤其是火山图中，我们可以看到，与MAST相比，edgeR识别出的上调基因比下调基因（在刺激与对照组之间）更多，而MAST识别出的上调和下调基因数量相似。

## 16.7. 主要结论

1. 在单细胞RNA测序中，来自同一实验对象的重复测量（细胞）会在测量值之间引入相关性，这需要通过将实验单位建模为随机效应来进行调整，以减轻伪重复问题。通过求和或求平均聚合（伪批量方法 - 通过固定或随机效应项）或通过将个体视为随机效应（单细胞方法）来考虑它们，以减轻伪重复问题。
2. 通过探索性分析确定变异的主要轴，以构建适当的设计矩阵，更好地模拟数据中的变异。
3. 在实验设计中，通过增加更多的样本可以最大限度地提高统计效力。

## 16.8. 思考题

1. 什么是差异基因表达，我们在哪些情况下对其进行测试感兴趣？
2. 什么是“伪重复”问题，如何避免它？
3. 什么是“多重检验”问题，如何规避它？

## 16.9. References

[deBH95](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id23) Yoav Benjamini and Yosef Hochberg. Controlling the false discovery rate: a practical and powerful approach to multiple testing. *Journal of the Royal Statistical Society: Series B (Methodological)*, 57(1):289–300, 1995. URL: [https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x](https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x), [arXiv:https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1995.tb02031.x](https://arxiv.org/abs/https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1995.tb02031.x), [doi:https://doi.org/10.1111/j.2517-6161.1995.tb02031.x](https://doi.org/https://doi.org/10.1111/j.2517-6161.1995.tb02031.x).  [deBKvB+17](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id4) Mollie E. Brooks, Kasper Kristensen, Koen J. van Benthem, Arni Magnusson, Casper W. Berg, Anders Nielsen, Hans J. Skaug, Martin Mächler, and Benjamin M. Bolker. glmmTMB Balances Speed and Flexibility Among Packages for Zero-inflated Generalized Linear Mixed Modeling. *The R Journal*, 9(2):378–400, 2017. URL: [https://doi.org/10.32614/RJ-2017-066](https://doi.org/10.32614/RJ-2017-066), [doi:10.32614/RJ-2017-066](https://doi.org/10.32614/RJ-2017-066).  deDRM+21([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id5),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id7)) Samarendra Das, Anil Rai, Michael L Merchant, Matthew C Cave, and Shesh N Rai. A comprehensive survey of statistical approaches for differential expression analysis in single-cell term`RNA` sequencing studies. *Genes (Basel)*, 12(12):1947, December 2021.  deFMY+15([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id3),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id20)) Greg Finak, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex K. Shalek, Chloe K. Slichter, Hannah W. Miller, M. Juliana McElrath, Martin Prlic, Peter S. Linsley, and Raphael Gottardo. Mast: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell term`rna` sequencing data. *Genome Biology*, 16(1):278, Dec 2015. URL: [https://doi.org/10.1186/s13059-015-0844-5](https://doi.org/10.1186/s13059-015-0844-5), [doi:10.1186/s13059-015-0844-5](https://doi.org/10.1186/s13059-015-0844-5).  [deHHAN+21](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id11) Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. *Cell*, 2021. URL: [https://doi.org/10.1016/j.cell.2021.04.048](https://doi.org/10.1016/j.cell.2021.04.048), [doi:10.1016/j.cell.2021.04.048](https://doi.org/10.1016/j.cell.2021.04.048).  [deHTTI17](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id6) Stephanie C Hicks, F William Townes, Mingxiang Teng, and Rafael A Irizarry. Missing data and technical variability in single-cell term`RNA`-sequencing experiments. *Biostatistics*, 19(4):562–578, 11 2017. URL: [https://doi.org/10.1093/biostatistics/kxx053](https://doi.org/10.1093/biostatistics/kxx053), [arXiv:https://academic.oup.com/biostatistics/article-pdf/19/4/562/26346801/kxx053.pdf](https://arxiv.org/abs/https://academic.oup.com/biostatistics/article-pdf/19/4/562/26346801/kxx053.pdf), [doi:10.1093/biostatistics/kxx053](https://doi.org/10.1093/biostatistics/kxx053).  [deJSeyetermDNAsrollahME16](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id7) Maria K Jaakkola, Fatemeh Seyeterm`DNA`srollah, Arfa Mehmood, and Laura L Elo. Comparison of methods to detect differentially expressed genes between single-cell populations. *Briefings in Bioinformatics*, 18(5):735–743, 07 2016. URL: [https://doi.org/10.1093/bib/bbw057](https://doi.org/10.1093/bib/bbw057), [arXiv:https://academic.oup.com/bib/article-pdf/18/5/735/25581122/bbw057.pdf](https://arxiv.org/abs/https://academic.oup.com/bib/article-pdf/18/5/735/25581122/bbw057.pdf), [doi:10.1093/bib/bbw057](https://doi.org/10.1093/bib/bbw057).  deJSE22([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id8),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id12)) Sini Junttila, Johannes Smolander, and Laura L Elo. Benchmarking methods for detecting differential states between conditions from multi-subject single-cell term`rna`-seq data. *bioRxiv*, 2022. URL: [https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662](https://www.biorxiv.org/content/early/2022/02/19/2022.02.16.480662), [doi:10.1101/2022.02.16.480662](https://doi.org/10.1101/2022.02.16.480662).  [deKST+18](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id14) Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell term`rna`-sequencing using natural genetic variation. *Nature biotechnology*, 36(1):89–94, 2018.  [deLZD+20](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id16) Charity W. Law, Kathleen Zeglinski, Xueyi Dong, Monther Alhamdoosh, Gordon K. Smyth, and Matthew E. Ritchie. A guide to creating design matrices for gene expression experiments. *F1000Research*, 9:1444–1444, Dec 2020. 33604029[pmid]. URL: [https://pubmed.ncbi.nlm.nih.gov/33604029](https://pubmed.ncbi.nlm.nih.gov/33604029).  [deLHA14](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id2) Michael I. Love, Wolfgang Huber, and Simon Anders. Moderated estimation of fold change and dispersion for term`rna`-seq data with deseq2. *Genome Biology*, 15(12):550, Dec 2014. URL: [https://doi.org/10.1186/s13059-014-0550-8](https://doi.org/10.1186/s13059-014-0550-8), [doi:10.1186/s13059-014-0550-8](https://doi.org/10.1186/s13059-014-0550-8).  deLT19([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id6),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id22)) Malte D Luecken and Fabian J Theis. Current best practices in single-cell term`rna`-seq analysis: a tutorial. *Molecular Systems Biology*, 15(6):e8746, 2019. URL: [https://www.embopress.org/doi/abs/10.15252/msb.20188746](https://www.embopress.org/doi/abs/10.15252/msb.20188746), [arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746](https://arxiv.org/abs/https://www.embopress.org/doi/pdf/10.15252/msb.20188746), [doi:https://doi.org/10.15252/msb.20188746](https://doi.org/https://doi.org/10.15252/msb.20188746).  [deLahnemannKosterS+20](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id18) David Lähnemann, Johannes Köster, Ewa Szczurek, Davis J. McCarthy, Stephanie C. Hicks, Mark D. Robinson, Catalina A. Vallejos, Kieran R. Campbell, Niko Beerenwinkel, Ahmed Mahfouz, Luca Pinello, Pavel Skums, Alexandros Stamatakis, Camille Stephan-Otto Attolini, Samuel Aparicio, Jasmijn Baaijens, Marleen Balvert, Buys de Barbanson, Antonio Cappuccio, Giacomo Corleone, Bas E. Dutilh, Maria Florescu, Victor Guryev, Rens Holmer, Katharina Jahn, Thamar Jessurun Lobo, Emma M. Keizer, Indu Khatri, Szymon M. Kielbasa, Jan O. Korbel, Alexey M. Kozlov, Tzu-Hao Kuo, Boudewijn P.F. Lelieveldt, Ion I. Mandoiu, John C. Marioni, Tobias Marschall, Felix Mölder, Amir Niknejad, Lukasz Raczkowski, Marcel Reinders, Jeroen de Ridder, Antoine-Emmanuel Saliba, Antonios Somarakis, Oliver Stegle, Fabian J. Theis, Huan Yang, Alex Zelikovsky, Alice C. McHardy, Benjamin J. Raphael, Sohrab P. Shah, and Alexander Schönhuth. Eleven grand challenges in single-cell data science. *Genome Biology*, 21(1):31, Feb 2020. URL: [https://doi.org/10.1186/s13059-020-1926-6](https://doi.org/10.1186/s13059-020-1926-6), [doi:10.1186/s13059-020-1926-6](https://doi.org/10.1186/s13059-020-1926-6).  [deMS22](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id13) Alan E. Murphy and Nathan G. Skene. A balanced measure shows superior performance of pseudobulk methods in single-cell RNA-sequencing analysis. *Nat Commun*, dec 2022. URL: [https://doi.org/10.1038%2Fs41467-022-35519-4](https://doi.org/10.1038%2Fs41467-022-35519-4), [doi:10.1038/s41467-022-35519-4](https://doi.org/10.1038/s41467-022-35519-4).  [deRPW+15](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id10) Matthew E. Ritchie, Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, and Gordon K. Smyth. Limma powers differential expression analyses for term`rna`-sequencing and microarray studies. *Nucleic acids research*, 43(7):e47–e47, Apr 2015. gkv007[PII]. URL: [https://doi.org/10.1093/nar/gkv007](https://doi.org/10.1093/nar/gkv007), [doi:10.1093/nar/gkv007](https://doi.org/10.1093/nar/gkv007).  deRMS10([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id1),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id15)) Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth. Edger: a bioconductor package for differential expression analysis of digital gene expression data. *Bioinformatics (Oxford, England)*, 26(1):139–140, Jan 2010. btp616[PII]. URL: [https://doi.org/10.1093/bioinformatics/btp616](https://doi.org/10.1093/bioinformatics/btp616), [doi:10.1093/bioinformatics/btp616](https://doi.org/10.1093/bioinformatics/btp616).  [deSR18](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id7) Charlotte Soneson and Mark D. Robinson. Bias, robustness and scalability in single-cell differential expression analysis. *Nature Methods*, 15(4):255–261, Apr 2018. URL: [https://doi.org/10.1038/nmeth.4612](https://doi.org/10.1038/nmeth.4612), [doi:10.1038/nmeth.4612](https://doi.org/10.1038/nmeth.4612).  deSGK+21([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id7),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id8)) Jordan W. Squair, Matthieu Gautier, Claudia Kathe, Mark A. Anderson, Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, Taha Qaiser, Kaya J. E. Matson, Quentin Barraud, Ariel J. Levine, Gioele La Manno, Michael A. Skinnider, and Grégoire Courtine. Confronting false discoveries in single-cell differential expression. *Nature Communications*, 12(1):5692, Sep 2021. URL: [https://doi.org/10.1038/s41467-021-25960-2](https://doi.org/10.1038/s41467-021-25960-2), [doi:10.1038/s41467-021-25960-2](https://doi.org/10.1038/s41467-021-25960-2).  [deTRCP21](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id17) Andrew L Thurman, Jason A Ratcliff, Michael S Chimenti, and Alejandro A Pezzulo. Differential gene expression analysis for multi-subject single cell term`RNA` sequencing studies with aggregateBioVar. *Bioinformatics*, 37(19):3243–3251, May 2021.  [deVRS+17](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id6) Catalina A. Vallejos, Davide Risso, Antonio Scialdone, Sandrine Dudoit, and John C. Marioni. Normalizing single-cell term`rna` sequencing data: challenges and opportunities. *Nature Methods*, 14(6):565–571, Jun 2017. URL: [https://doi.org/10.1038/nmeth.4292](https://doi.org/10.1038/nmeth.4292), [doi:10.1038/nmeth.4292](https://doi.org/10.1038/nmeth.4292).  [deWLNN19](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id5) Tianyu Wang, Boyang Li, Craig E. Nelson, and Sheida Nabavi. Comparative analysis of differential gene expression analysis tools for single-cell term`rna` sequencing data. *BMC Bioinformatics*, 20(1):40, Jan 2019. URL: [https://doi.org/10.1186/s12859-019-2599-6](https://doi.org/10.1186/s12859-019-2599-6), [doi:10.1186/s12859-019-2599-6](https://doi.org/10.1186/s12859-019-2599-6).  deZEL21([1](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id8),[2](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id9),[3](https://www.sc-best-practices.org/conditions/differential_gene_expression.html#id19)) Kip D. Zimmerman, Mark A. Espeland, and Carl D. Langefeld. A practical solution to pseudoreplication bias in single-cell studies. *Nature Communications*, 12(1):738, Feb 2021. URL: [https://doi.org/10.1038/s41467-021-21038-1](https://doi.org/10.1038/s41467-021-21038-1), [doi:10.1038/s41467-021-21038-1](https://doi.org/10.1038/s41467-021-21038-1). 

## 16.10. Contributors

We gratefully acknowledge the contributions of:

### 16.10.1. Authors

- Lukas Heumos
- Anastasia Litinetskaya
- Soroor Hediyeh-Zadeh

### 16.10.2. Reviewers
