
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../rnasequp_1/">
      
      
        <link rel="next" href="../../RNASEQ/rnaseq_1/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.2">
    
    
      
        <title>1-2. 上游分析全教程 - Bioinformatic Tutorial (Daxing Research)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7bf56d0a.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#rna-seq" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-header__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatic Tutorial (Daxing Research)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1-2. 上游分析全教程
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
            </label>
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-nav__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Bioinformatic Tutorial (Daxing Research)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        写在前面
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          基础环境配置
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          基础环境配置
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_1/" class="md-nav__link">
        配置1. Python环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_2/" class="md-nav__link">
        配置2. Linux环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_3/" class="md-nav__link">
        配置3. R语言环境
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
      
      
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          RNA-seq上游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq上游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../rnasequp_1/" class="md-nav__link">
        1-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          1-2. 上游分析全教程
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        1-2. 上游分析全教程
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 测序数据下载
  </a>
  
    <nav class="md-nav" aria-label="1. 测序数据下载">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-sra" class="md-nav__link">
    1.1 SRA文件下载
  </a>
  
    <nav class="md-nav" aria-label="1.1 SRA文件下载">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#111-sra" class="md-nav__link">
    1.1.1 单SRA文件下载
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#112-sra" class="md-nav__link">
    1.1.2 多SRA文件下载
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-srafastq" class="md-nav__link">
    1.2 sra格式转fastq格式
  </a>
  
    <nav class="md-nav" aria-label="1.2 sra格式转fastq格式">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#121" class="md-nav__link">
    1.2.1 单端测序文件转换
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#122" class="md-nav__link">
    1.2.2 双端测序文件转换
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2" class="md-nav__link">
    2. 测序数据质控
  </a>
  
    <nav class="md-nav" aria-label="2. 测序数据质控">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21" class="md-nav__link">
    2.1 检测数据质量
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#211" class="md-nav__link">
    2.1.1 单份报告生成
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#212" class="md-nav__link">
    2.1.2 多份报告整合
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22" class="md-nav__link">
    2.2 测序数据过滤
  </a>
  
    <nav class="md-nav" aria-label="2.2 测序数据过滤">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#221" class="md-nav__link">
    2.2.1 单端测序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#222" class="md-nav__link">
    2.2.2 双端测序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#223" class="md-nav__link">
    2.2.3 多组测序数据同时过滤
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3" class="md-nav__link">
    3. 比对到参考基因组
  </a>
  
    <nav class="md-nav" aria-label="3. 比对到参考基因组">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31" class="md-nav__link">
    3.1 参考基因组文件下载
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-subread" class="md-nav__link">
    3.2 subread比对
  </a>
  
    <nav class="md-nav" aria-label="3.2 subread比对">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#321" class="md-nav__link">
    3.2.1 构建索引
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#322" class="md-nav__link">
    3.2.2 比对
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-counts" class="md-nav__link">
    4. 统计基因counts数
  </a>
  
    <nav class="md-nav" aria-label="4. 统计基因counts数">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-gtf" class="md-nav__link">
    4.1 下载gtf基因组注释文件
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-featurecounts" class="md-nav__link">
    4.2 使用featureCounts进行定量分析
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          RNA-seq下游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq下游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_1/" class="md-nav__link">
        2-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_2/" class="md-nav__link">
        2-2. 差异表达基因(DEG)分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_3/" class="md-nav__link">
        2-3. 通路富集分析(GSEA)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          单细胞样本对齐
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          单细胞样本对齐
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_1/" class="md-nav__link">
        3-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_2/" class="md-nav__link">
        3-2. 数据预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_3/" class="md-nav__link">
        3-3. 模型准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_4/" class="md-nav__link">
        3-4. 模型训练
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_5/" class="md-nav__link">
        3-5. 对齐细胞
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          MOFA单细胞多组学因子分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          MOFA单细胞多组学因子分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_1/" class="md-nav__link">
        4-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_2/" class="md-nav__link">
        4-2. 单细胞样本对齐
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_3/" class="md-nav__link">
        4-3. scRNA-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_4/" class="md-nav__link">
        4-4. scATAC-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_5/" class="md-nav__link">
        4-5. MOFA模型构建
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_6/" class="md-nav__link">
        4-6. MOFA下游分析-R语言环节
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_7/" class="md-nav__link">
        4-7. MOFA下游分析-Python环节
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          TMT蛋白组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          TMT蛋白组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_1/" class="md-nav__link">
        5-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_2/" class="md-nav__link">
        5-2. TMT蛋白组学上游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_3/" class="md-nav__link">
        5-3. TMT蛋白组学下游分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          多组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          多组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-1/" class="md-nav__link">
        6-1. 数据准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-2/" class="md-nav__link">
        6-2. GLUE多组学整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-3/" class="md-nav__link">
        6-3. 细胞亚群分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-4/" class="md-nav__link">
        6-4. 差异表达分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-5/" class="md-nav__link">
        6-5. 细胞通讯分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-6/" class="md-nav__link">
        6-6. 动态调控分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-7/" class="md-nav__link">
        6-7. 转录因子网络
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          Python可视化
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Python可视化
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_1/" class="md-nav__link">
        1-1. 火山图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_2/" class="md-nav__link">
        1-2. 箱线图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_3/" class="md-nav__link">
        1-3. 热图(聚类)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          视网膜母细胞瘤分析教程
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          视网膜母细胞瘤分析教程
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/rb_tutorial/" class="md-nav__link">
        Base
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_1_preprocess/" class="md-nav__link">
        教程1: 单细胞上游结果预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_2_filtered_qc/" class="md-nav__link">
        教程2: 质量控制与过滤
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_3_glue/" class="md-nav__link">
        教程3: scRNA-seq与scATAC-seq整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_4_scbasset/" class="md-nav__link">
        教程4: Motif活性分数计算
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_5_analysis1/" class="md-nav__link">
        教程5: 综合分析1
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_6_analysis2/" class="md-nav__link">
        教程6: 综合分析1-Chromatin activity和Motif score评估
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_7_STER_cell/" class="md-nav__link">
        教程7: STER cell下游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_8_kineic_photo/" class="md-nav__link">
        教程8: 转录动力学分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_9_cellphonedb/" class="md-nav__link">
        教程9: 细胞通讯分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 测序数据下载
  </a>
  
    <nav class="md-nav" aria-label="1. 测序数据下载">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-sra" class="md-nav__link">
    1.1 SRA文件下载
  </a>
  
    <nav class="md-nav" aria-label="1.1 SRA文件下载">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#111-sra" class="md-nav__link">
    1.1.1 单SRA文件下载
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#112-sra" class="md-nav__link">
    1.1.2 多SRA文件下载
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-srafastq" class="md-nav__link">
    1.2 sra格式转fastq格式
  </a>
  
    <nav class="md-nav" aria-label="1.2 sra格式转fastq格式">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#121" class="md-nav__link">
    1.2.1 单端测序文件转换
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#122" class="md-nav__link">
    1.2.2 双端测序文件转换
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2" class="md-nav__link">
    2. 测序数据质控
  </a>
  
    <nav class="md-nav" aria-label="2. 测序数据质控">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21" class="md-nav__link">
    2.1 检测数据质量
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#211" class="md-nav__link">
    2.1.1 单份报告生成
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#212" class="md-nav__link">
    2.1.2 多份报告整合
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22" class="md-nav__link">
    2.2 测序数据过滤
  </a>
  
    <nav class="md-nav" aria-label="2.2 测序数据过滤">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#221" class="md-nav__link">
    2.2.1 单端测序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#222" class="md-nav__link">
    2.2.2 双端测序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#223" class="md-nav__link">
    2.2.3 多组测序数据同时过滤
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3" class="md-nav__link">
    3. 比对到参考基因组
  </a>
  
    <nav class="md-nav" aria-label="3. 比对到参考基因组">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31" class="md-nav__link">
    3.1 参考基因组文件下载
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-subread" class="md-nav__link">
    3.2 subread比对
  </a>
  
    <nav class="md-nav" aria-label="3.2 subread比对">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#321" class="md-nav__link">
    3.2.1 构建索引
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#322" class="md-nav__link">
    3.2.2 比对
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-counts" class="md-nav__link">
    4. 统计基因counts数
  </a>
  
    <nav class="md-nav" aria-label="4. 统计基因counts数">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-gtf" class="md-nav__link">
    4.1 下载gtf基因组注释文件
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-featurecounts" class="md-nav__link">
    4.2 使用featureCounts进行定量分析
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


<h1 id="rna-seq">RNA-seq. 上游分析全教程</h1>
<h2 id="1">1. 测序数据下载</h2>
<p>在本节中，我们将介绍如何从ncbi上下载原始测序数据SRA文件进行分析</p>
<h3 id="11-sra">1.1 SRA文件下载</h3>
<h4 id="111-sra">1.1.1 单SRA文件下载</h4>
<p>如果你只想分析一个SRA文件，并且已经知道了SRA号，那么我们可以在linux交互环境输入以下命令进行下载</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#单SRA下载prefetch</span>
<span class="linenos" data-linenos="2 "></span><span class="n">SRR10502962</span>
</code></pre></div>
<h4 id="112-sra">1.1.2 多SRA文件下载</h4>
<p>如果你需要批量下载SRA文件，那么你需要得到一个带有多个SRA号的txt文件</p>
<p>e.g. SRR_Acc_List.txt文件内容</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">SRR5113012</span>
<span class="linenos" data-linenos="2 "></span><span class="n">SRR5113013</span>
<span class="linenos" data-linenos="3 "></span><span class="n">SRR5113014</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#多SRA下载</span>
<span class="linenos" data-linenos="2 "></span><span class="n">prefetch</span> <span class="o">--</span><span class="n">option</span><span class="o">-</span><span class="n">file</span> <span class="n">SRR_Acc_List</span><span class="o">.</span><span class="n">txt</span>
</code></pre></div>
<h3 id="12-srafastq">1.2 sra格式转fastq格式</h3>
<p>sra格式的文件一般是经过压缩的测序文件，我们需要转换成原始测序数据fastq格式。</p>
<h4 id="121">1.2.1 单端测序文件转换</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#转换当前目录下全部以.sra结尾的文件</span>
<span class="linenos" data-linenos="2 "></span><span class="n">fastq</span><span class="o">-</span><span class="n">dump</span> <span class="o">*.</span><span class="n">sra</span>
</code></pre></div>
<h4 id="122">1.2.2 双端测序文件转换</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#--split-files参数可以将其分解为两个fastq文件。</span>
<span class="linenos" data-linenos="2 "></span><span class="n">fastq</span><span class="o">-</span><span class="n">dump</span> <span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">files</span> <span class="o">*.</span><span class="n">sra</span>
</code></pre></div>
<h2 id="2">2. 测序数据质控</h2>
<p>为了检测我们测序数据的质量，我们常常需要生成一份质控报告进行直观的观察</p>
<h4 id="21">2.1 检测数据质量</h4>
<h4 id="211">2.1.1 单份报告生成</h4>
<p>在本小节，我们使用fastqc生成质控报告，在fastq文件目录下输入</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#批量生成所有fastq文件的质控报告</span>
<span class="linenos" data-linenos="2 "></span><span class="n">fastqc</span> <span class="o">*.</span><span class="n">fastq</span>
</code></pre></div>
<p>等待运行结束后，在同目录下有着<em>_fastqc.html和</em>_fastqc.zip两个文件，我们可以打开对应的html文件查看该fastq数据的质量。报告的解读见文章</p>
<h4 id="212">2.1.2 多份报告整合</h4>
<p>通过上面的步骤，我们得到了每个fastq文件的质控报告，为了整体进行评估，我们使用multiqc整合报告结果</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#整合质控报告结果</span>
<span class="linenos" data-linenos="2 "></span><span class="n">multiqc</span> <span class="o">*.</span><span class="n">zip</span>
</code></pre></div>
<h3 id="22">2.2 测序数据过滤</h3>
<p>由于我们得到的数据可能包括接头序列，引物，poly-A尾巴和其他类型的不需要的序列，为避免影响下面的分析，我们需要去除这些无关的测序数据。</p>
<p>在这里，相较于其他过滤工具而言，我们选择trim-galore</p>
<h4 id="221">2.2.1 单端测序</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1">#新建clean文件夹存放测序结果</span>
<span class="linenos" data-linenos=" 2 "></span>mkdir<span class="w"> </span>clean
<span class="linenos" data-linenos=" 3 "></span><span class="c1">#单端测序数据质量过滤</span>
<span class="linenos" data-linenos=" 4 "></span>trim_galore<span class="w"> </span>-q<span class="w"> </span><span class="m">20</span><span class="w"> </span><span class="se">\ </span><span class="c1">#设定Phred quality score阈值</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">    </span>--phred33<span class="w"> </span><span class="se">\ </span><span class="c1">#选择-phred33或者-phred64，表示测序平台使用的Phred qualityscore</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">    </span>--stringency<span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="se">\ </span><span class="c1">#设定可以忍受的前后adapter重叠的碱基数，默认为1（非常苛刻）。可以适度放宽，因为后一个adapter几乎不可能被测序仪读到</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">    </span>--length<span class="w"> </span><span class="m">20</span><span class="w"> </span><span class="se">\ </span><span class="c1">#设定输出reads长度阈值，小于设定值会被抛弃。</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">    </span>-e<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\ </span><span class="c1">#容错率</span>
<span class="linenos" data-linenos=" 9 "></span><span class="w">    </span>-o<span class="w"> </span>/home/seq/clean<span class="w"> </span><span class="c1">#输出结果</span>
<span class="linenos" data-linenos="10 "></span><span class="w">    </span>/home/seq/SRR10502962_1.fastq<span class="w"> </span><span class="c1">#待处理文件</span>
</code></pre></div>
<h4 id="222">2.2.2 双端测序</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1">#新建clean文件夹存放测序结果</span>
<span class="linenos" data-linenos=" 2 "></span>mkdir<span class="w"> </span>clean
<span class="linenos" data-linenos=" 3 "></span><span class="c1">#双端测序数据质量过滤</span>
<span class="linenos" data-linenos=" 4 "></span>trim_galore<span class="w"> </span>-q<span class="w"> </span><span class="m">25</span><span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">    </span>--phred33<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">    </span>--stringency<span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">    </span>--length<span class="w"> </span><span class="m">36</span><span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">    </span>-e<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos=" 9 "></span><span class="w">    </span>--paired<span class="w"> </span><span class="c1">#双端测序</span>
<span class="linenos" data-linenos="10 "></span><span class="w">    </span>-o<span class="w"> </span>/home/seq/clean
<span class="linenos" data-linenos="11 "></span><span class="w">    </span>/home/seq/SRR10502962_1.fastq<span class="w"> </span>/home/seq/SRR10502962_2.fastq
</code></pre></div>
<h4 id="223">2.2.3 多组测序数据同时过滤</h4>
<p>为了同时完成多组测序数据的同时过滤，我们在这里编写.sh脚本来在Linux系统下批量运行</p>
<p><strong>config</strong></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>mkdir<span class="w"> </span>clean
<span class="linenos" data-linenos="2 "></span><span class="nb">cd</span><span class="w"> </span>clean
<span class="linenos" data-linenos="3 "></span>ls<span class="w"> </span>/home/seq/*_1.fastq<span class="w"> </span>&gt;1
<span class="linenos" data-linenos="4 "></span>ls<span class="w"> </span>/home/seq/*_2.fastq<span class="w"> </span>&gt;2
<span class="linenos" data-linenos="5 "></span>paste<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="m">2</span><span class="w">  </span>&gt;<span class="w"> </span>config
</code></pre></div>
<p><strong>qc.sh</strong></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nv">bin_trim_galore</span><span class="o">=</span>trim_galore
<span class="linenos" data-linenos="2 "></span><span class="nv">dir</span><span class="o">=</span><span class="s1">&#39;/home/seq/clean&#39;</span>
<span class="linenos" data-linenos="3 "></span>cat<span class="w"> </span><span class="nv">$1</span><span class="w"> </span><span class="p">|</span><span class="k">while</span><span class="w"> </span><span class="nb">read</span><span class="w"> </span>id
<span class="linenos" data-linenos="4 "></span><span class="k">do</span>
<span class="linenos" data-linenos="5 "></span><span class="w">    </span>arr<span class="o">(</span><span class="si">${</span><span class="nv">id</span><span class="si">}</span><span class="o">)</span>
<span class="linenos" data-linenos="6 "></span><span class="w">    </span><span class="nv">fq1</span><span class="o">=</span><span class="si">${</span><span class="nv">arr</span><span class="p">[0]</span><span class="si">}</span>
<span class="linenos" data-linenos="7 "></span><span class="w">    </span><span class="nv">fq2</span><span class="o">=</span><span class="si">${</span><span class="nv">arr</span><span class="p">[1]</span><span class="si">}</span>
<span class="linenos" data-linenos="8 "></span><span class="nv">$bin_trim_galore</span><span class="w"> </span>-q<span class="w"> </span><span class="m">25</span><span class="w"> </span>--phred33<span class="w"> </span>--phred33<span class="w"> </span>--length<span class="w"> </span><span class="m">36</span><span class="w"> </span>--stringency<span class="w"> </span><span class="m">3</span><span class="w"> </span>--paired<span class="w"> </span>-o<span class="w"> </span><span class="nv">$dir</span><span class="w"> </span><span class="nv">$fq1</span><span class="w"> </span><span class="nv">$fq2</span><span class="w"> </span>
<span class="linenos" data-linenos="9 "></span><span class="k">done</span>
</code></pre></div>
<p>运行qc.sh</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#config是传递进去的参数</span>
<span class="linenos" data-linenos="2 "></span>bash<span class="w"> </span>qc.sh<span class="w"> </span>config
</code></pre></div>
<h2 id="3">3. 比对到参考基因组</h2>
<p>由于测序仪机器读长的限制，在构建文库的过程中首先需要将DNA片段化，测序得到的序列只是基因组上的部分序列。为了确定测序reads在基因组上的位置，需要将reads比对回参考基因组上，这个步骤叫做mapping</p>
<h3 id="31">3.1 参考基因组文件下载</h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#参考基因组下载(hg38)(subread)</span>
<span class="linenos" data-linenos="2 "></span>wget<span class="w"> </span>-O<span class="w">  </span>hg38.fa.gz<span class="w"> </span>http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
<span class="linenos" data-linenos="3 "></span><span class="c1">#解压</span>
<span class="linenos" data-linenos="4 "></span>gunzip<span class="w"> </span>hg38.fa.gz
</code></pre></div>
<h3 id="32-subread">3.2 subread比对</h3>
<h4 id="321">3.2.1 构建索引</h4>
<p>subread有着极其快速的比对效率，这与其构建索引的预处理是不可分开的，所以我们用以下代码来构建索引</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nb">cd</span><span class="w"> </span>/home/seq/index/hg38
<span class="linenos" data-linenos="2 "></span>subread-buildindex<span class="w"> </span>-o<span class="w"> </span>hg38<span class="w"> </span>hg38
</code></pre></div>
<h4 id="322">3.2.2 比对</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>subread-align<span class="w"> </span>-t<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\ </span><span class="c1">#0代表RNA-seq，1代表DNA-seq</span>
<span class="linenos" data-linenos="2 "></span><span class="w">    </span>-T5<span class="w"> </span><span class="se">\ </span><span class="c1">#线程数</span>
<span class="linenos" data-linenos="3 "></span><span class="w">    </span>-i<span class="w"> </span>hg38<span class="w">  </span><span class="se">\ </span><span class="c1">#指定参考基因组的basename</span>
<span class="linenos" data-linenos="4 "></span><span class="w">    </span>-r<span class="w"> </span>/home/seq/SRR10502962_1.fastq<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="5 "></span><span class="w">    </span>-R<span class="w"> </span>/home/seq/SRR10502962_2.fastq<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="6 "></span><span class="w">    </span>-o<span class="w"> </span>SRR10502962.bam<span class="w"> </span><span class="c1">#输出文件</span>
</code></pre></div>
<h2 id="4-counts">4. 统计基因counts数</h2>
<p>在这里，我们仅介绍一个工具featureCounts。</p>
<p>featuresCounts软件用于统计基因/转录本上mapping的reads数，也就是用于raw count定量。该软件不仅支持基因/转录本的定量，也支持exon，gene bodies，genomic bins，chromsomal locations等区间的定量</p>
<h3 id="41-gtf">4.1 下载gtf基因组注释文件</h3>
<p>撰写本教程时的最新版本为103，如有更新可以去官网看看再来下载</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#我们将基因组注释文件存放到refer文件夹中</span>
<span class="linenos" data-linenos="2 "></span>mkdir<span class="w"> </span>/home/seq/refer
<span class="linenos" data-linenos="3 "></span><span class="nb">cd</span><span class="w"> </span>/home/seq/refer
<span class="linenos" data-linenos="4 "></span><span class="c1">#下载</span>
<span class="linenos" data-linenos="5 "></span>wget<span class="w"> </span>http://ftp.ensembl.org/pub/release-103/gtf/homo_sapiens/Homo_sapiens.GRCh38.103.gtf.gz
</code></pre></div>
<h3 id="42-featurecounts">4.2 使用featureCounts进行定量分析</h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="c1">#选择gtf路径</span>
<span class="linenos" data-linenos=" 2 "></span><span class="nv">gtf</span><span class="o">=</span><span class="s2">&quot;/home/seq/refer/Homo_sapiens.GRCh38.103.gtf.gz&quot;</span>
<span class="linenos" data-linenos=" 3 "></span><span class="c1">#featureCounts 定量分析</span>
<span class="linenos" data-linenos=" 4 "></span>featureCounts<span class="w"> </span>-T<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\ </span><span class="c1">#线程数</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">    </span>-p<span class="w"> </span><span class="se">\ </span><span class="c1">#针对paired-end数据</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">    </span>-t<span class="w"> </span>exon<span class="w"> </span><span class="se">\ </span><span class="c1">#跟-g一样的意思，其是默认将exon作为一个feature</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">    </span>-g<span class="w"> </span>gene_id<span class="w"> </span><span class="se">\ </span><span class="c1">#从注释文件中提取Meta-features信息用于read count，默认是gene_id</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">    </span>-a<span class="w"> </span><span class="nv">$gtf</span><span class="w"> </span><span class="se">\ </span><span class="c1">#输入GTF/GFF基因组注释文件</span>
<span class="linenos" data-linenos=" 9 "></span><span class="w">    </span>-o<span class="w"> </span>all.id.txt<span class="w"> </span><span class="se">\ </span><span class="c1">#输出文件</span>
<span class="linenos" data-linenos="10 "></span><span class="w">    </span>*.bam<span class="w"> </span><span class="c1">#待处理数据</span>
<span class="linenos" data-linenos="11 "></span>
<span class="linenos" data-linenos="12 "></span><span class="c1">#去除多余信息，矩阵保存为counts.txt</span>
<span class="linenos" data-linenos="13 "></span>cat<span class="w"> </span>all.id.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut-f1,7-<span class="w"> </span>&gt;<span class="w"> </span>counts.txt
</code></pre></div>
<p>到这里，我们就已经得到一个Counts矩阵了，后续的分析被称为转录组下游分析。将在后面的教程中继续介绍</p>
<table>
<thead>
<tr>
<th>Geneid</th>
<th>SRR10502962.bam</th>
</tr>
</thead>
<tbody>
<tr>
<td>ENSG00000223972</td>
<td>1</td>
</tr>
<tr>
<td>ENSG00000227232</td>
<td>132</td>
</tr>
<tr>
<td>ENSG00000278267</td>
<td>3</td>
</tr>
<tr>
<td>ENSG00000243485</td>
<td>0</td>
</tr>
</tbody>
</table>





                

  

              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fc8c2696.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>